{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c964bb8a-c503-47cb-908d-d976553ee482",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Company Introduction of Wecomput\n",
      "\n",
      "Wecomput™ is a High-Tech company focusing on using computing technology to drive the development of innovative drugs research and benefit human health.\n",
      "\n",
      "The core team members come from world-renowned AI pharmaceutical companies, leading pharmaceutical companies, well-known Internet companies. They have rich experience, deep understanding and enthusiasm in cross-cutting fields such as pharmaceuticals, life sciences, artificial intelligence, and software development.\n",
      "\n",
      "Wecomput Technology Co., Ltd., was registered and established in 2015. R&D centers were set in Beijing, Shanghai and Guangzhou. We obtained equity investment from Sequoia in 2021 and were selected into the NVIDIA Inception program. So far, Wecomput have served hundreds of customers all over the world, covering many colleges, hospitals, institutes, and biopharmas. Deep cooperation has been established between some of them. Besides, we have published hundreds of SCI papers as well as the invention patents and software copyrights.\n",
      "\n",
      "Wecomput developed WeMol, a digital intelligent computing platform which is designed for empowering the whole process of drug development. It supports digital computing and AI modeling of proteins, antibodies, nucleic acids, chemicals, biological materials, and natural products. WeMol adopts a streaming architecture, which supports low-code development and flexible expansion. The core algorithm modules exceed or are comparable to international mainstream commercial software in terms of speed, accuracy, storage efficiency and other indicators\n",
      "\n",
      "\n",
      "# WeMol Modules\n",
      "\n",
      "## 1. 2D Similarity Search\n",
      "\n",
      "The 2D Similarity Search module is a tool based on molecular fingerprints for 2D similarity search. Firstly, the fingerprint bit-vector or vector string of the template small molecule is calculated based on the fingerprint types (Maccs Key, pharmacophore ﬁngerprints, extended connectivity fingerprints). Then, the fingerprint bit-vector or vector string is used for molecular similarity search in the selected public library or private library, and the small molecules that are similar (or dissimilar) to the template molecule are obtained.\n",
      "\n",
      "## 2. 3-letter AA Conversion\n",
      "\n",
      "Convert 3-letter amino acids to 1-letter amino acid. E.g., \"\"ASP ILE VAL ASN\"\" will be converted to -> \"\"DIVQ\"\".\n",
      "\n",
      "## 3. 3D Conf Generation (AlphaConf)\n",
      "\n",
      "AlphaConf is a super fast 3D conformation search and generation engine. Machine learning models for bond lengths/angles based on millions of high-quality data in PubChemQC. A new way of defining restriction structure fragments is developed to generate the three-dimensional conformation of molecules, and the accuracy is better than similar algorithms. By adopting the non-repetitive conformation generation method, a lot of computation time is saved, and the computation speed is much faster than similar algorithms. The unique and efficient conformation compression technology reduces the storage space by 400-800 times compared with similar algorithms and is suitable for the construction of ultra-large-scale 3D conformation libraries and ultra-high-throughput virtual screening.\n",
      "\n",
      "## 4. AA Probability Prediction\n",
      "\n",
      "Leveraging pre-trained large-scale protein language models, otherwise known as PLMs or pLLMs, there emerges the capacity to forecast the likelihood of each of the twenty amino acids appearing at any given position within a sequence. Comparable to the structurally conservative amino acids found in evolution, those with high probability predictions from the language model are beneficial in enhancing the protein's stability, fostering more efficient protein folding, augmenting its expression capacity, and potentially elevating its affinity. By comparison, this implies an intrinsic advantage over haphazardly induced mutations.\n",
      "\n",
      "## 5. AC2SDF\n",
      "\n",
      "The AC2SDF module is used to convert the compressed binary conformation file AC.GZ generated by the AlphaConf module into an SDF file for easier viewing.\n",
      "\n",
      "## 6. ADMET Prediction\n",
      "\n",
      "ADMET Prediction based on Communicative Message Passing Neural Network. CMPNN improves molecular embedding by strengthening the message interactions between nodes and edges through a communicative kernel. In addition, the message generation process is enriched by introducing a new message booster module.\n",
      "\n",
      "## 7. ADMET Prediction (v2)\n",
      "\n",
      "ADMET Prediction (v2) is a machine learning-based module for predicting the ADMET properties of small molecules. It enables rapid batch predictions of ADMET properties and supports four common and efficient machine learning algorithms: Graph Attention Neural Network (GAT), Light Gradient Boosting Machine (LightGBM), Random Forest, and Gradient Boosting Machine (GBM). The module supports two methods for molecular feature representation: molecular fingerprints and molecular descriptors, allowing for quick batch predictions on libraries of small molecule compounds. It supports 27 ADMET properties, including 7 regression models and 20 classification models. Module select the ideal machine learning algorithm and molecular characterization method based on the predictive performance data provided in the documentation.\n",
      "\n",
      "## 8. Absolute Folding Stability\n",
      "\n",
      "The Absolute Folding Stability Prediction module predicts the absolute stability ΔG of proteins using the protein sequence generation model ESM-IF. When tested on collected datasets, the model exhibited a prediction error with an RMSE of approximately 1.5 kcal/mol and a Pearson correlation coefficient of 0.7. This advancement reduces the cost of obtaining protein stability ΔG through experimental methods such as CD and fluorescence quantitative PCR, making a significant breakthrough in computationally predicting the folding stability ΔG of proteins.\n",
      "\n",
      "## 9. Aggregation Score\n",
      "\n",
      "Aggregation Scores is designed for the rational design of protein solubility and aggregation tendency in protein structures. It allows researchers to predict aggregation-prone sites in proteins by considering both sequence and structure. This is particularly useful for globular proteins, where aggregation-prone sites may be buried within the native structure and the sequence may be discontinuous. By considering experimental aggregation propensity scales of natural amino acids, this method can accurately predict protein aggregation tendency.\n",
      "\n",
      "## 10. Alanine Scan\n",
      "\n",
      "Alanine scanning involves mutating each residue of a protein to Ala and calculating the resulting change in free energy. It is useful for rapid scanning because polar interactions and steric hindrance within the residue will be disrupted upon mutation to Ala.\n",
      "\n",
      "## 11. Alanine Scan (MMPBSA v2)\n",
      "\n",
      "Alanine Scan (MMPBSA) calculates components of binding free energy after alanine mutation using the MM-PBSA method.\n",
      "\n",
      "## 12. Alanine Scan (MMPBSA)\n",
      "\n",
      "Alanine Scan (MMPBSA) calculates components of binding free energy after alanine mutation using the MM-PBSA method.\n",
      "\n",
      "## 13. AlphaAutoMD (GMX2023 v2)\n",
      "\n",
      "The AlphaAutoMD module is a fully automated molecular dynamics module, which enables fast submit an MD job using a PDB file. This module integrates MD PDB Prepare, Protein Protonation, GMX Receptor Parameterization, GMX Ligand Parameterization, MD Solvation, GMX MD Run, and RMS.\n",
      "\n",
      "## 14. AlphaAutoMD (GMX2023)\n",
      "\n",
      "The AlphaAutoMD module is a fully automated molecular dynamics module, which enables fast submit an MD job using a PDB file. This module integrates MD PDB Prepare, Protein Protonation, GMX Receptor Parameterization, GMX Ligand Parameterization, MD Solvation, GMX MD Run, and RMS.\n",
      "\n",
      "## 15. AlphaShape\n",
      "\n",
      "AlphaShape is a molecular shape and pharmacophore-based virtual screening module. The AlphaShape algorithm performs virtual screening or protein structure search based on the three-dimensional similarity of molecules on the basis of three-dimensional conformation. By combining the Gaussian function and the deep neural network model, the calculation accuracy achieves SOTA.\n",
      "\n",
      "## 16. Antibody (Off-) Target Prediction\n",
      "\n",
      "Structure similarity-based antibody (Off-) target prediction from antibody-antigen interaction database.\n",
      "\n",
      "## 17. Antibody Design (MEAN)\n",
      "\n",
      "Antibody Design (MEAN) module is implemented based on the MEAN model, which utilizes a multi-channel equivariant graph attention network. It can be used to design the one-dimensional sequence and three-dimensional structure of CDRs.\n",
      "\n",
      "## 18. Antibody Fv Humanness\n",
      "\n",
      "Score humanness for antibody Fv sequence using deep learning models. The Humanness Score is an indexed number from 0 (non-human) to 1 (human), trained using billions of antibody Fv sequences from NGS studies.\n",
      "\n",
      "## 19. Antibody Fv Structure Prediction (IgFold)\n",
      "\n",
      "IgFold, a fast deep learning method for antibody structure prediction. IgFold consists of a pre-trained language model trained on 558M natural antibody sequences followed by graph networks that directly predict backbone atom coordinates. IgFold predicts structures of similar or better quality than alternative methods (including AlphaFold) in significantly less time (under one minute).\n",
      "\n",
      "## 20. Antibody Mutation Analysis\n",
      "\n",
      "Antibody Mutation Analysis is a DeepAb-based novel mutant scoring metric derived from network confidence, the top-ranked mutations improve binding affinity. This model will be helpful for a broad range of antibody prediction and design tasks.\n",
      "\n",
      "## 21. Antibody Numbering\n",
      "\n",
      "Antibody Numbering is a module for antibody numbering for variable regions and constant regions. Mainstream numbering schemes are supported, e.g., Kabat, Chothia, and IMGT are widely used for Fv, and EU is the most used scheme for the constant region.\n",
      "\n",
      "## 22. Antibody Numbering v2\n",
      "\n",
      "Antibody Numbering v2 is a module for antibody numbering for variable regions and constant regions. Mainstream numbering schemes are supported, e.g., Kabat, Chothia, and IMGT are widely used for Fv, and EU is the most used scheme for the constant region.\n",
      "\n",
      "## 23. Antibody Paratope Prediction\n",
      "\n",
      "Predict the amino acid sites on an antibody that bind to an antigen, known as the Paratope. The algorithm is based on a deep learning model using an isomorphic graph neural network, trained and predicted on antibody structures, and has the highest prediction accuracy among existing methods.\n",
      "\n",
      "## 24. Antibody RMSD\n",
      "\n",
      "The Antibody RMSD module performs a CDR region-based structure superposition of the reference antibody and other CDR identical antibody structures, and calculates the RMSD value of the CDR region.\n",
      "\n",
      "## 25. Antibody Sequence Prediction (AbLang)\n",
      "\n",
      "Language models trained on the antibody sequences in the OAS database have the power of predicting or restoring missing residues in antibody sequences.\n",
      "\n",
      "## 26. Antibody Sequence Prediction (IgLM)\n",
      "\n",
      "The Antibody Sequence Prediction (IgLM) module is designed for the generation and optimization of antibody sequences, utilizing the Observed Antibody Space (OAS) to collect antibody sequences. The OAS database contains natural antibody sequences from six species: humans, mice, rats, rabbits, rhesus monkeys, and camels. To investigate the impact of model capacity, two versions of the model were trained: IgLM and IgLM-S, with 13 million and 1.4 million training parameters, respectively. Both IgLM models were trained on 5.58 million non-redundant sequences, clustered based on 95% similarity. During training, 10 to 20 residues in the antibody sequences were randomly masked to enable diversification of arbitrary spans during inference. Additionally, constraints were applied to the chain type (heavy chain or light chain) and the originating species of the sequences, providing a framework to control the generation of species-specific antibody sequences. This method has been shown to produce full-length heavy and light chain sequences from various species, as well as improved developability for filling CDR loop libraries. The method serves as a powerful tool for antibody design and can be applied to various antibody sequence design scenarios.\n",
      "\n",
      "## 27. Antibody Viscosity Prediction\n",
      "\n",
      "Viscosity is an important factor affecting the development of antibody drugs. Clinically, antibodies often need to be administered intravenously or subcutaneously, requiring a high concentration of antibody solution (>100mg/mL) to obtain a therapeutic dose at a small dose. However, high concentrations of antibodies often exhibit high viscosity, which poses a challenge to the development, manufacture and administration of antibody drugs. It has been found that antibody sequence is the key factor to determine antibody viscosity. It has been reported that antibody viscosity is correlated with charge in Fv region, charge asymmetry in VH and VL region, FvCSP, and hydrophobic index HI in Fv region. It is an effective method to predict antibody viscosity based on antibody sequence.\n",
      "\n",
      "## 28. Antibody-Antigen Docking (HADDOCK)\n",
      "\n",
      "HADDOCK v3.0 is a bottom-up reimagination of the long standing time-proven HADDOCK used for integrative modeling of biomolecular complexes. This young and still very experimental (use it at your own risk!) aims to modularize and extend HADDOCK’s core functions. In its current implementation, HADDOCK v3.0 still lacks the full repertoire of features present at the production web server version, HADDOCK v2.4. However it is able to take full advantage of the ambiguous interaction restraints (AIRs) to drive the docking process.\n",
      "\n",
      "## 29. Back Mutation Grouping\n",
      "\n",
      "Back Mutation Grouping is a grouping module in the antibody humanization design workflow, which groups the back mutations based on the mutation scoring table generated by the Mutation Score module and returns the back mutated sequence.\n",
      "\n",
      "## 30. Back Mutation Grouping (v2.3)\n",
      "\n",
      "Back Mutation Grouping is a grouping module in the antibody humanization design workflow, which groups the back mutations based on the mutation scoring table generated by the Mutation Score module and returns the back mutated sequence.\n",
      "\n",
      "## 31. Back Mutation Grouping (v2.4)\n",
      "\n",
      "Back Mutation Grouping is a grouping module in the antibody humanization design workflow, which groups the back mutations based on the mutation scoring table generated by the Mutation Score module and returns the back mutated sequence.\n",
      "\n",
      "## 32. Batch Renaming\n",
      "\n",
      "Batch Renaming module was designed to molecule rename for chemical library. User could standardize the molecule name using prefix and defined length. For example, rename a library start from WCP0001 to WCP9999, user could input prefix with WCP and length of 4. User also could preserve the previous name using --keeptitle parameter to preserve the relations between names. This module could use for customized molecules naming in a large de novo library or a user private chemical library.\n",
      "\n",
      "## 33. CIF2PDB\n",
      "\n",
      "The CIF2PDB module is based on BioPython to convert mmCIF files into PDB files.\n",
      "\n",
      "## 34. Cleavage Site Prediction\n",
      "\n",
      "Predict protein cleavage sites for eight commonly used proteases (trypsin, ArgC, chymotrypsin, GluC, LysC, AspN, LysN, and LysargiNase).\n",
      "\n",
      "## 35. Codon Optimization\n",
      "\n",
      "Codon optimization can be used for optimizing codons (i.e., the genetic code) for the automated oligonucleotide design of gene synthesis based on PCR. The availability of whole genome sequences has greatly increased the number of protein targets, many of which need to be overexpressed in cells outside of their native DNA source. Synthetic genes can be optimized for expression and constructed to be easily mutagenized without consideration for the parental genome. However, designing and constructing synthetic genes, particularly those encoding large proteins, can be a slow, difficult, and confusing process. This module automatically designs oligonucleotides for gene synthesis using a PCR-based approach.\n",
      "\n",
      "## 36. Complex Interface Analysis\n",
      "\n",
      "Complex Interface Analysis module is designed for identifying the residues on the interface of a structure complex.\n",
      "\n",
      "## 37. Cyclic Peptide Design\n",
      "\n",
      "The Cyclic Peptide Design module uses the cyclic peptide design algorithm AfCycDesign developed by the research groups of Sergey Ovchinnikov at Harvard University and Gaurav Bhardwaj at the University of Washington. Cyclic Peptide Design uses AlphaFold modified network for accurate structure prediction and design of cyclic peptides. Results show this approach can accurately predict the structures of native cyclic peptides from a single sequence, with 36 out of 49 cases predicted with high confidence (pLDDT > 0.85) matching the native structure with root mean squared deviation (RMSD) less than 1.5 Å. Redesign of macrocyclic backbone sequences using AlphaFold resulted in seven sequences with different sizes and structures that matched the X-ray crystal structures very closely.\n",
      "\n",
      "## 38. Cyclic Peptide Structure Prediction\n",
      "\n",
      "Cyclic Peptide Structure Prediction module uses the sequence of linear peptides to generate the structure of a cyclic peptide.\n",
      "\n",
      "## 39. De novo Generation\n",
      "\n",
      "REINVENT is a method to tune a sequence-based generative model for molecular de novo design that through augmented episodic likelihood can learn to generate structures with certain specified desirable properties. We demonstrate how this model can execute a range of tasks such as generating analogues to a query structure and generating compounds predicted to be active against a biological target. As a proof of principle, the model is first trained to generate molecules that do not contain sulphur. As a second example, the model is trained to generate analogues to the drug Celecoxib, a technique that could be used for scaffold hopping or library expansion starting from a single molecule. Finally, when tuning the model towards generating compounds predicted to be active against the dopamine receptor type 2, the model generates structures of which more than 95% are predicted to be active, including experimentally confirmed actives that have not been included in either the generative model nor the activity prediction model.\n",
      "\n",
      "## 40. De novo Generation (Moses)\n",
      "\n",
      "De novo Generation (Moses) is a deep learning-based molecular generation module, which implements various mainstream molecular generation models, including character-level recurrent neural networks, variational autoencoders, and adversarial autoencoders.\n",
      "\n",
      "## 41. De novo Generation (REINVENT4)\n",
      "\n",
      "Small molecule de novo generation based on REINVENT4. REINVENT 4 enables and facilitates de novo design, R-group replacement, library design, linker design, scaffold hopping and molecule optimization.\n",
      "\n",
      "## 42. De-immunization Design\n",
      "\n",
      "Introduce progressively increasing mutations on sequences with immunogenicity to reduce or eliminate immunogenicity. This module is usually not used alone and is generally used in combination with AlphaMHC. For automated de-immunization, please use the De-immunization Design process.\n",
      "\n",
      "## 43. Descriptors (RDKit)\n",
      "\n",
      "Calculate 2D/3D molecular descriptors for molecules and write them out to a SD or CSV/TSV text file. 2D descriptors: Autocorr2D, MolWt, Ipc, NumRotatableBonds, qed, etc. 3D descriptors: Autocorr3D, RadiusOfGyration, Eccentricity, etc. (See the documentation for more descriptors). And FragmentCountOnly descriptors: fr_Al_COO, fr_Al_OH, fr_Al_OH_noTert, etc.\n",
      "\n",
      "## 44. Disulfide Bond Search\n",
      "\n",
      "Disulfide Bond Search module calculates potential disulfide bond locations in proteins, which is useful for optimizing protein stability. Disulfide bond plays a great role in the stability of protein, but the addition of unreasonable disulfide bond will easily lead to aggregation, expression reduction and even misfolding.\n",
      "\n",
      "## 45. Diverse Subset\n",
      "\n",
      "Pick a subset of diverse molecules based on a variety of 2D fingerprints using MaxMin or an available hierarchical clustering methodology and write them to a file. The Dice and Tanimoto similarity functions available in RDKit are able to handle fingerprints corresponding to both IntVect and BitVect. All other similarity functions, however, expect BitVect fingerprints to calculate pairwise similarity. Consequently, ExplicitBitVect fingerprints are generated for AtomPairs, Morgan, MorganFeatures, and TopologicalTorsions for similarity calculations instead of default IntVect fingerprints.\n",
      "\n",
      "## 46. Duplicates Removal\n",
      "\n",
      "Duplicates Removal modules identify and remove duplicate molecules based on canonical SMILES strings or simply count the number of duplicate molecules.\n",
      "\n",
      "## 47. Electrostatic Potential Calculation (APBS)\n",
      "\n",
      "Calculate the electrostatic potential energy of biomolecular structures using the APBS method and generate surface plots. To visualize the surface maps, execute this function from the structure editor WeView: WeView->Analysis->Electrostatics.\n",
      "\n",
      "## 48. Enumerate Stereoisomers\n",
      "\n",
      "Enumerate Stereoisomers is a tool for performing a combinatorial enumeration of stereoisomers for molecules around all or unassigned chiral atoms and bonds. cis-trans isomer and optical isomer are supported.\n",
      "\n",
      "## 49. Evaluate Nucleic Acid (AlphaRNA)\n",
      "\n",
      "The Evaluate Nucleic Acid (AlphaRNA) module is used to evaluate the expression and half-life of nucleic acid sequences, antibody titers, etc. Support human, mouse, rat, pig and other species.\n",
      "\n",
      "## 50. Extract Fv Sequence\n",
      "\n",
      "Extract Fv Sequence is a tool for Extracting the Fv region sequence from antibody full-length sequence.\n",
      "\n",
      "## 51. Extract Sequence from Structure (PDB2FASTA)\n",
      "\n",
      "Extracts the protein sequences in a PDB file to FASTA. Amino acids are represented by their one-letter code while all others are represented by 'X'.\n",
      "\n",
      "## 52. FASTA File\n",
      "\n",
      "FASTA File is a module for specifying fasta file which could used for other modules input.\n",
      "\n",
      "## 53. File\n",
      "\n",
      "File is a module for specifying file path which could be used for multiple modules.\n",
      "\n",
      "## 54. Format Conversion (Open Babel)\n",
      "\n",
      "Open Babel is a chemical toolbox designed to speak the many languages of chemical data. It's an open, collaborative project allowing anyone to search, convert, analyze, or store data from molecular modeling, chemistry, solid-state materials, biochemistry, or related areas.\n",
      "\n",
      "## 55. Format Conversion (RDKit)\n",
      "\n",
      "Format Conversion (RDKit) is a molecular file format conversion tool based on RDKit.\n",
      "\n",
      "## 56. Free Energy Surface Analysis\n",
      "\n",
      "Free energy surface analysis for PLUMED based metadynamics.\n",
      "\n",
      "## 57. GMX Ligand Parameterization\n",
      "\n",
      "Generate mol2, gro, and itp files for ligand in molecular dynamics using Gromacs.\n",
      "\n",
      "## 58. GMX MD Run (GMX2023)\n",
      "\n",
      "GMX MD Run (GMX2023) runs a Gromacs MD task using the prepared system topology and parameter files.\n",
      "\n",
      "## 59. GMX MDP Generation (Auto)\n",
      "\n",
      "The GMX MDP Generation (Auto) module is mainly based on the selected system (membrane, receptor, ligand) to automatically generate the MDP file required for the molecular dynamics simulation process. This file is the input file required for the Gromacs molecular dynamics simulation, which contains various parameters. To set more detailed parameters, go to the Minimize MDP Generation, NPT MDP Generation, MD MDP Generation module.\n",
      "\n",
      "## 60. GMX Membrane Parameterization\n",
      "\n",
      "The GMX Membrane Parameterization module is the GRO, ITP and TOP file that generates the membrane structure according to Amber or Charmm.\n",
      "\n",
      "## 61. GMX Metadynamics Generation\n",
      "\n",
      "The GMX Metadynamics Generation module is used to generate input files for Metadynamics simulations.\n",
      "\n",
      "## 62. GMX Receptor Parameterization\n",
      "\n",
      "Generate gro, itp, and top files for receptor (protein or nucleic acid) for molecular dynamics using Gromacs.\n",
      "\n",
      "## 63. Gene Enrichment (DAVID)\n",
      "\n",
      "Gene Enrichment (DAVID) is a gene function enrichment analysis module based on DAVID. DAVID is a bioinformatics database that integrates biological data and analysis tools, providing systematic and comprehensive biological functional annotation information for large-scale gene or protein lists.\n",
      "\n",
      "## 64. Generate Humanized Variants\n",
      "\n",
      "Generate humanized variant sequences based on the Grafting and Back Mutation Grouping results.\n",
      "\n",
      "## 65. Germline AA Distribution Frequency\n",
      "\n",
      "Germline AA Distribution Frequency module outputs the amino acid frequency distribution of the germline at each position of the antibody. It can output the distribution separately according to the specified germline gene family (usually focusing on the frequency distribution of the germline genes in the same family as the target sequence).\n",
      "\n",
      "## 66. Germline Blast\n",
      "\n",
      "Based on IgBlastp, the Germline Blast module will search for the homologous template closest to the target antibody sequence in the IMGT reference sequences database through sequence alignment and output the corresponding template sequence and sequence consistency. The default sequence types searched in the database are: IMGT V genes (F+ORF+in-frame P).\n",
      "\n",
      "## 67. Grafting (v2)\n",
      "\n",
      "Graft antibody CDRs to target frameworks, normally for humanization.\n",
      "\n",
      "## 68. Grafting (v2.1)\n",
      "\n",
      "Graft antibody CDRs to target frameworks, normally for humanization. Version: v2.1\n",
      "\n",
      "## 69. Grafting (v2.2)\n",
      "\n",
      "Graft antibody CDRs to target frameworks, normally for humanization. Version: v2.2\n",
      "\n",
      "## 70. Grafting (v2.3)\n",
      "\n",
      "Graft antibody CDRs to target frameworks, normally for humanization. Version: v2.3\n",
      "\n",
      "## 71. Grafting (v2.4)\n",
      "\n",
      "Graft antibody CDRs to target frameworks, normally for humanization. Version: v2.4\n",
      "\n",
      "## 72. Homology Modeling (Protein)\n",
      "\n",
      "Homology or comparative modeling of protein three-dimensional structures. Users provide a sequence to be modeled and compare it with known related structures. Protein structure modeling is performed by satisfying spatial constraint conditions, as well as many other tasks, including novel modeling of loops in protein structures, optimization of various protein structure models for flexibly defined objective functions, multiple alignments of protein sequences and/or structures, clustering, searching sequence databases, and comparing protein structures.\n",
      "\n",
      "## 73. Homology Tree\n",
      "\n",
      "The Homology Tree module is used to generate homologous evolutionary trees.\n",
      "\n",
      "## 74. Human Antibody BLAST\n",
      "\n",
      "BLAST human antibody database for homologs, which currently contains over one billion sequences, from over 75 different studies. These repertoires cover diverse immune states and individuals from humans. Submit an antibody sequence, and homologous human antibody sequences will be returned and could be used for advanced antibody humanization, affinity maturation, de-immunization, etc.\n",
      "\n",
      "## 75. Human Fragment BLAST\n",
      "\n",
      "The Human Fragment BLAST is based on inputs of 9 peptides, searching the Germline, TCR, NextProt, OAS for the most similar 9 peptides.\n",
      "\n",
      "## 76. Human Germline BLAST\n",
      "\n",
      "Search the human germline database for homologs of the target antibody sequence, and output the template sequences and the corresponding identities.\n",
      "\n",
      "## 77. Human Germline BLAST (v2.0)\n",
      "\n",
      "Search the human germline database for homologs of the target antibody sequence, and output the template sequences and the corresponding identities.\n",
      "\n",
      "## 78. Humanization Report\n",
      "\n",
      "Humanization Report is an antibody humanization design reporting module for Generating the humanization design reports as well as patent example paragraphs.\n",
      "\n",
      "## 79. Humanization Report (v2.3)\n",
      "\n",
      "Humanization Report is an antibody humanization design reporting module for Generating the humanization design reports as well as patent example paragraphs.\n",
      "\n",
      "## 80. Humanization Report (v2.4)\n",
      "\n",
      "Humanization Report is an antibody humanization design reporting module for Generating the humanization design reports as well as patent example paragraphs. Compared with v2.3, RMSD and energy information are added.\n",
      "\n",
      "## 81. IgG Modeling\n",
      "\n",
      "IgG Modeling module performs modeling on the full-length sequence of antibodies to construct the complete three-dimensional structure of IgG, supporting both monospecific and bispecific antibodies. It automatically identifies the variable region (Fv) sequences within the full-length sequence and models them using state-of-the-art methods (currently ESMFold). The remaining parts of the IgG, including the Fc and linker, are modeled using homology modeling based on the crystal structures of known full-length antibodies as templates, with spatial constraints. This approach yields better results than directly predicting the complete IgG structure using methods like AF2.\n",
      "\n",
      "## 82. Immune Protein Structure Prediction\n",
      "\n",
      "Immune Protein Structure Prediction module is based on ImmuneBuilder. ImmuneBuilder is a set of deep learning models that accurately predict the structure of antibodies (ABodyBuilder2), NanoBodyBuilder2, and T-cell receptors (TCRBuilder2). ImmuneBuilder generates structures with state-of-the-art precision while being much faster than AlphaFold2.\n",
      "\n",
      "## 83. Immunogenicity Prediction (AlphaMHC v2.0)\n",
      "\n",
      "The AlphaMHC algorithm utilizes popular NLP natural language processing technology and a novel multimodal fusion deep neural network architecture. It integrates nearly one billion publicly and privately available wet lab experimental data related to immunogenicity (including affinity data, NGS data, mass spectrometry data, etc.) for training. It successfully achieves end-to-end prediction of immunogenicity risk from sequence to clinical application and has been validated using over a hundred clinical real-world immunogenicity data from FDA and EMA (including mono-/multi-specific antibodies and recombinant proteins). AlphaMHC can accurately distinguish between high and low immunogenicity, with an ROC-AUC of 0.87 and an accuracy of over 80% (up to 91% for some test sets). It exhibits significantly superior predictive performance compared to existing methods and is the only algorithm known to have been validated with clinical data.\n",
      "\n",
      "## 84. Immunogenicity Prediction (AlphaMHC v3.0 beta)\n",
      "\n",
      "The AlphaMHC algorithm employs popular NLP (Natural Language Processing) techniques and a novel multi-modal fusion deep neural network architecture. It integrates nearly one billion wet lab data points related to immunogenicity (including affinity data, NGS data, mass spectrometry data, etc.) for training, achieving end-to-end prediction from sequence to clinical immunogenicity risk. This has been validated with hundreds of real clinical immunogenicity data points from the FDA and EMA (including mono- and multi-specific antibodies, recombinant proteins, etc.). This version is the latest and has been tested primarily with clinical ADA data from antibodies, achieving an accuracy of 90% and an AUC of 0.91. Its performance surpasses that of version 2.0, and it is recommended for trial.\n",
      "\n",
      "## 85. Immunogenicity Prediction (WeADApt v4.1.0)\n",
      "\n",
      "Internal test version of WeComput independently developed latest immunogenicity prediction algorithm: AlphaMHC v4.1.0 beta.\n",
      "\n",
      "## 86. MD (GMX2024)\n",
      "\n",
      "GMX MD Run (GMX2024) runs a Gromacs MD task using the prepared system topology and parameter files.\n",
      "\n",
      "## 87. MD Clustering\n",
      "\n",
      "MD Clustering is a clustering analysis of dynamic trajectories.\n",
      "\n",
      "## 88. MD Clustering (v2)\n",
      "\n",
      "MD Clustering is a clustering analysis of dynamic trajectories.\n",
      "\n",
      "## 89. MD Distance\n",
      "\n",
      "MD Distance is a distance analysis module that outputs the distance changes between two groups (center of mass distance or geometric center distance) over time.\n",
      "\n",
      "## 90. MD Distance (v2)\n",
      "\n",
      "MD Distance is a distance analysis module that outputs the distance changes between two groups (center of mass distance or geometric center distance) over time.\n",
      "\n",
      "## 91. MD Gyration\n",
      "\n",
      "MD Gyration cycloidal radius analysis, which can be used to measure the average radius of pledge during system simulation.\n",
      "\n",
      "## 92. MD Gyration (v2)\n",
      "\n",
      "MD Gyration cycloidal radius analysis, which can be used to measure the average radius of pledge during system simulation.\n",
      "\n",
      "## 93. MD Hbond\n",
      "\n",
      "MD Hbond for hydrogen bond analysis between specified groups.\n",
      "\n",
      "## 94. MD Hbond (v2)\n",
      "\n",
      "MD Hbond for hydrogen bond analysis between specified groups.\n",
      "\n",
      "## 95. MD MDP Generation\n",
      "\n",
      "Generate final Gromacs MD production MDP file.\n",
      "\n",
      "## 96. MD PCA\n",
      "\n",
      "MD PCA (Principal component analysis) module can analyze the main influencing factors (eigenvectors) from the high-dimensional data. The first few eigenvectors (principal components, such as PC1 and PC2 for the first two principal components) can generally describe most of the information about molecular motion. The motion path of a flexible large system with N atoms, such as protein, needs 3N Cartesian coordinates to describe, so it is difficult to understand and intuitively analyze the high-dimensional data.\n",
      "\n",
      "## 97. MD PCA (v2)\n",
      "\n",
      "MD PCA (Principal component analysis) module can analyze the main influencing factors (eigenvectors) from the high-dimensional data. The first few eigenvectors (principal components, such as PC1 and PC2 for the first two principal components) can generally describe most of the information about molecular motion. The motion path of a flexible large system with N atoms, such as protein, needs 3N Cartesian coordinates to describe, so it is difficult to understand and intuitively analyze the high-dimensional data.\n",
      "\n",
      "## 98. MD PDB Prepare\n",
      "\n",
      "MD PDB Prepare is a structure preparation module before running molecular dynamics. The missing residues in PDB were added using PDBFixer. The protein, nucleic acid, and ligands were extracted and output individually.\n",
      "\n",
      "## 99. MD RMS\n",
      "\n",
      "The RMS module calculates the RMSD or RMSF to analyze the structural stability of the system.\n",
      "\n",
      "## 100. MD RMS (v2)\n",
      "\n",
      "The RMS module calculates the RMSD or RMSF to analyze the structural stability of the system.\n",
      "\n",
      "## 101. MD SASA\n",
      "\n",
      "MD SASA module calculates the solvent accessible surface area (SASA) for a specified group.\n",
      "\n",
      "## 102. MD SASA (v2)\n",
      "\n",
      "MD SASA module calculates the solvent accessible surface area (SASA) for a specified group.\n",
      "\n",
      "## 103. MD Solvation\n",
      "\n",
      "MD Solvation module adds water box and ions for the system.\n",
      "\n",
      "## 104. MD Solvation v2\n",
      "\n",
      "Adds water box and ions for the system. Add user-specified ions in version v2.\n",
      "\n",
      "## 105. MD Trajectory\n",
      "\n",
      "MD Trajectory converts Gromacs trajectory file (xtc) into GRO or PDB file for visualization.\n",
      "\n",
      "## 106. MD Trajectory (v2)\n",
      "\n",
      "MD Trajectory converts Gromacs trajectory file (xtc) into GRO or PDB file for visualization.\n",
      "\n",
      "## 107. MHC-I Binding Prediction\n",
      "\n",
      "MHC-I Binding Prediction is a neural network-based model for predicting MHC-I binding affinity. The model is trained using affinity (BA) and mass spectrometry eluted ligand (MS eluted ligand) data, and it incorporates the prediction of the affinity values and peptide lengths of specific MHC molecules using the NNAlign framework. The NetMHCpan-4.0 method improves the accuracy of predicting tumor neoantigens, validated eluted ligands (ELs), and T-cell epitopes.\n",
      "\n",
      "## 108. MMPBSA\n",
      "\n",
      "MMPBSA calculates the binding free energy between the receptor and ligand and provides energy decomposition data, binding constant (Ka), and inhibitor constant (Ki).\n",
      "\n",
      "## 109. MMPBSA (v2)\n",
      "\n",
      "MMPBSA calculates the binding free energy between the receptor and ligand and provides energy decomposition data, binding constant (Ka), and inhibitor constant (Ki).\n",
      "\n",
      "## 110. Membrane Solvation\n",
      "\n",
      "Membrane Solvation module adds water box and ions for the membrane, receptor, ligand.\n",
      "\n",
      "## 111. Membrane System Construction\n",
      "\n",
      "Membrane System Construction module builds the PDB file of the membrane structure.\n",
      "\n",
      "## 112. Metabolism Site Prediction\n",
      "\n",
      "Metabolism Site Prediction module predict which sites in a molecule are most liable to metabolism by Cytochrome P450. The supported small-molecule input file formats are SD (.sdf,.sd) and SMILES (.smi).\n",
      "\n",
      "## 113. Minimize MDP Generation\n",
      "\n",
      "The Minimize MDP Generation module is primarily used to generate input MDP files that are required for Minimization of Gromacs molecular dynamics simulations.\n",
      "\n",
      "## 114. Molecular Docking (AutoDock-GPU)\n",
      "\n",
      "The Molecular Docking (AutoDock-GPU) module is a docking simulation tool used primarily to predict binding modes and interactions between molecules and obtain information such as molecular docking energy and binding affinity. It can also calculate and compare the binding abilities of multiple molecules, making it useful for drug molecule screening, design, and optimization. AutoDock-GPU is the OpenCL and Cuda accelerated version of AutoDock4.2.6, which leverages its embarrassingly parallelizable LGA by processing ligand-receptor poses in parallel over multiple compute units. The supported input formats for ligand files are SD (.sdf, .sd), PDB (.pdb), and MOL (.mol). The supported input format for receptor files is PDB (.pdb).\n",
      "\n",
      "## 115. Molecular Docking (DiffDock)\n",
      "\n",
      "Molecular Docking is a diffusion-forming model that has been used for docking between small molecules and proteins. DiffDock achieved a top-1 success rate of 38% (RMSD<2A) on PDBBind, significantly exceeding the previous state-of-the-art of traditional docking (23%) and deep learning (20%) approaches. In addition, while previous methods were unable to butt computationally folded structures (with a maximum accuracy of 10.4%), DiffDock maintained a significantly higher accuracy (21.7%). Finally, DiffDock has fast reasoning times and provides confidence estimates with high selective precision.\n",
      "\n",
      "## 116. Molecule In Membrane\n",
      "\n",
      "The Molecule In Membrane module is a structural file that generates receptor/ligand/complex orientation.\n",
      "\n",
      "## 117. Multiple Sequence Alignment\n",
      "\n",
      "Align multiple sequences using progressive alignment algorithm for evolutionary analysis, generating phylogenetic trees.\n",
      "\n",
      "## 118. Multiple Sequence Alignment (MAFFT)\n",
      "\n",
      "mafft - Multiple alignment program for amino acid or nucleotide sequences\n",
      "\n",
      "## 119. Mutation Energy of Binding\n",
      "\n",
      "Mutation Energy of Binding module aims to calculate the effect of mutations on the binding energy of a complex. Based on the input complex structure and mutation files, it builds mutation structures and calculates the binding energy between chains. By comparing with the wild type, it calculates the change in binding energy between chains before and after the mutation. The more negative the energy, the more favorable the mutation is for binding between specified chains.\n",
      "\n",
      "## 120. Mutation Energy of Binding (GeoPPI)\n",
      "\n",
      "Deep geometric representations for modeling effects of mutations on protein-protein binding affinity.\n",
      "\n",
      "## 121. Mutation Energy of Binding (Pythia-PPI)\n",
      "\n",
      "Deep learning and multi-task learning based prediction of protein-protein binding affinity changes upon mutations.\n",
      "\n",
      "## 122. Mutation Energy of Stability\n",
      "\n",
      "Mutation Energy of Stability is used to predict the effect of mutations on protein stability. First, the protein structure is energy optimized, and one or more residues are mutated to new ones. Protein stability analysis is then performed by calculating the change in folding free energy.\n",
      "\n",
      "## 123. Mutation Energy of Stability (Pythia)\n",
      "\n",
      "A self-supervised graph neural network for protein stability prediction upon mutation.\n",
      "\n",
      "## 124. Mutation Energy of Stability (ThermoMPNN)\n",
      "\n",
      "The Mutation Energy of Stability (ThermoMPNN) module is implemented based on the ThermoMPNN model, a deep neural network model that predicts the stability changes corresponding to a single point mutation based on the initial structure of the protein. The model uses structural features extracted from ProteinMPNN, a deep neural network model that predicts a protein's amino acid sequence based on its three-dimensional structure, to achieve excellent prediction performance on established baseline datasets.\n",
      "\n",
      "## 125. Mutation Format Conversion\n",
      "\n",
      "Mutation Format Conversion is a tool to convert the list of mutations applicable to the sequence to a format suitable for the structure by adding the chain name.\n",
      "\n",
      "## 126. Mutation List Generation\n",
      "\n",
      "Generate a list of mutations (aka. consensus mutations) from a set of aligned sequences (normally generated by the blast).\n",
      "\n",
      "## 127. Mutation Score (v2.0)\n",
      "\n",
      "Mutation Score is a core module in antibody humanization design workflow, which is a structure-based automated scoring module. Based on the structure information of the antibody and the CDR-grafted sequence information, this module quantitatively scores the degree of change before and after the replacement of each amino acid in the FR region. The higher the score, the greater the potential impact of the amino acid replacement on the conformation change of the CDR region during CDR grafting, indicating the need for auto-back mutation. The module outputs the score for each amino acid, which is used for subsequent grouping and generation of humanized antibody sequences in the antibody humanization design workflow.\n",
      "\n",
      "## 128. Mutation Score (v2.1)\n",
      "\n",
      "Mutation Score is a core module in antibody humanization design workflow, which is a structure-based automated scoring module. Based on the structure information of the antibody and the CDR-grafted sequence information, this module quantitatively scores the degree of change before and after the replacement of each amino acid in the FR region. The higher the score, the greater the potential impact of the amino acid replacement on the conformation change of the CDR region during CDR grafting, indicating the need for auto-back mutation. The module outputs the score for each amino acid, which is used for subsequent grouping and generation of humanized antibody sequences in the antibody humanization design workflow.\n",
      "\n",
      "## 129. Mutation Score (v2.3)\n",
      "\n",
      "Mutation Score is a core module in antibody humanization design workflow, which is a structure-based automated scoring module. Based on the structure information of the antibody and the CDR-grafted sequence information, this module quantitatively scores the degree of change before and after the replacement of each amino acid in the FR region. The higher the score, the greater the potential impact of the amino acid replacement on the conformation change of the CDR region during CDR grafting, indicating the need for auto-back mutation. The module outputs the score for each amino acid, which is used for subsequent grouping and generation of humanized antibody sequences in the antibody humanization design workflow.\n",
      "\n",
      "## 130. Mutation Score (v2.4)\n",
      "\n",
      "Mutation Score is a core module in antibody humanization design workflow, which is a structure-based automated scoring module. Based on the structure information of the antibody and the CDR-grafted sequence information, this module quantitatively scores the degree of change before and after the replacement of each amino acid in the FR region. The higher the score, the greater the potential impact of the amino acid replacement on the conformation change of the CDR region during CDR grafting, indicating the need for auto-back mutation. The module outputs the score for each amino acid, which is used for subsequent grouping and generation of humanized antibody sequences in the antibody humanization design workflow.\n",
      "\n",
      "## 131. NGS Analysis\n",
      "\n",
      "This module is used for DNA sequence (antibody) analysis after NGS sequencing: IGV, IGD, IGJ clonotype annotation; amino acid sequence translation; antibody numbering and CDR recognition; uncommon residues and high frequency mutations idenfication using protein (antibody) language models (ESM, IgLM); PTM hot-spot liability analysis; Sequence-based physico-chemical property calculation including pI (isoelectric point), molecular weight, hydrophobicity index; sequence clustering; SHM (somatic hyper-mutation) rate calculation, etc.\n",
      "\n",
      "## 132. NPT MDP Generation\n",
      "\n",
      "Generate Gromacs MD input file at constant temperature and pressure (NPT).\n",
      "\n",
      "## 133. Nanobody Humanization (Llamanade)\n",
      "\n",
      "Nanobody Humanization (Llamanade) module is implemented based on Llamanade, which systematically analyzes the sequence and structural properties of Nbs based on NGS databases and high-resolution structures. A large amount of framework diversity was revealed and key differences between Nbs and human immunoglobulin G (IgG) antibodies were highlighted. Conserved residues that may contribute to improved solubility, structural stability, and antigen-binding were identified to facilitate the rational humanization of Nbs. It has been successfully applied to humanize a group of structurally diverse and potent SARS-CoV-2 neutralized Nbs. It takes less than a minute to perform a comprehensive humanization analysis of a given Nbs.\n",
      "\n",
      "## 134. PAINS Filter\n",
      "\n",
      "Filter Pan-assay Interference molecules (PAINS) from an input file by performing a substructure search using SMARTS pattern and write out appropriate molecules to an output file or simply count the number of filtered molecules.\n",
      "\n",
      "## 135. PDB File\n",
      "\n",
      "PDB File is a module for specifying pdb file which could used for other modules input.\n",
      "\n",
      "## 136. PDB Fixer\n",
      "\n",
      "The PDB Fixer is a module that fixes problems in the PDB file. It can automatically fix the following problems: add missing heavy atoms; add missing hydrogen atoms; build missing loops; convert non-standard residues to their standard equivalents; select a single position for atoms with multiple alternate positions listed; delete unwanted chains from the model; delete unwanted heterogens; build a water box for explicit solvent simulations.\n",
      "\n",
      "## 137. PDB Insertion Removal\n",
      "\n",
      "Renumber the antibody PDB file to remove any insertion codes in UID, to make such PDB compatible with other tools.\n",
      "\n",
      "## 138. PDB Mutation\n",
      "\n",
      "Mutate a protein structure in PDB format and return mutated structure.\n",
      "\n",
      "## 139. PDB ReNumbering\n",
      "\n",
      "PDB ReNumbering is a tool module that renumbers protein residues and supports renumbering antibody structure with kabat, imgt, and chothia schemes. It takes a protein structure PDB file as input and outputs a renumbered PDB file.\n",
      "\n",
      "## 140. PTM Hotspot by Sequence\n",
      "\n",
      "Scan antibody sequences for potential PTM (post-translational modification) hotspots (liabilities). PTM hotspot is a common risk for biologics development. It is generally recommended to use the PTM function in WeSeq for visual analysis. This module is more commonly used for assembling automated workflows.\n",
      "\n",
      "## 141. PTM Hotspot by Structure\n",
      "\n",
      "Prediction of isomerization probability of aspartic acid (ASP) site in protein Structure by PTM Hotspot by Structure.\n",
      "\n",
      "## 142. Patch Analysis\n",
      "\n",
      "Patch typically refers to regions on the protein surface that are enriched with positively charged, negatively charged, or hydrophobic residues, which can potentially lead to protein-protein interactions or aggregation. Mutations can often be used to eliminate or reduce patches, thereby effectively improving the degree of aggregation or viscosity. This module analyzes the size and distribution of patches on the protein surface to address protein aggregation issues. It is recommended to use the WeView 3D structure visualization editor to utilize this feature, as it allows for a visual inspection of the patch locations within the 3D structure.\n",
      "\n",
      "## 143. Patent BLAST\n",
      "\n",
      "A module for sequence retrieval of antibody full-length or CDR region. When retrieving an antibody variable region from a patent, existing BLAST programs (such as NCBI BLAST) usually search the whole sequence, but for antibodies, the function mainly depends on the CDR, FR is relatively not important, and due to the generality of FR, FR of many different antibodies is the same or highly homologous. However, FR accounts for a higher proportion of sequences, resulting in a lot of sequences with similar FR but different CDR by BLAST in the variable region of antibodies. Moreover, in addition to protecting the complete sequence of the variable region during patent application, in many cases, the antibody CDR will also be protected separately to obtain a wider range of protection, so it is necessary to search for homologous sequences with CDR as the target in the process of antibody development. To this end, the Vixon team developed the program, which can retrieve the closest sequence to the target CDR from the existing patent library. Data updated: Dec, 2024\n",
      "\n",
      "## 144. Patent CDR BLAST\n",
      "\n",
      "Patent CDR BLAST is a module designed for sequence retrieval of the antibody CDR region. Existing BLAST programs (such as NCBI BLAST) typically search with the entire sequence when searching for an antibody's variable region from a patent. However, for antibodies, function mainly depends on the CDR, with the FR being relatively unimportant. Additionally, due to the universality of the FR, many different antibodies have the same or highly homologous FRs, which results in BLAST searches for the antibody's variable region returning many sequences that are similar in FR but dissimilar in CDR. Furthermore, when filing a patent application, in addition to protecting the sequence of the variable region, many instances also protect the antibody's CDR separately to obtain greater protection. Therefore, when developing antibodies, it is necessary to search for homologous sequences with the CDR as the target. To address this, the WECOMPUT team developed this module that can accurately retrieve sequences from existing patent databases closest to the target CDR and ignore the influence of the FR.\n",
      "\n",
      "## 145. Patent Sequence Listing\n",
      "\n",
      "Patent Sequence Listing is a tool for extracting sequences in bulk from patent text files. Many macromolecule patents come with a sequence listing file that contains all the sequences in the patent claims. However, it is difficult for humans to efficiently read and extract these sequences. With this module, all sequences can be extracted in bulk at once. The Image(OCR) is the conversion of image-based protein sequences into 3-letter coded or 1-letter coded sequences.\n",
      "\n",
      "## 146. Peptide Structure Generation\n",
      "\n",
      "Peptide Structure Generation is a tool for generating peptide structures based on peptide sequences. Input the amino acid sequence of the peptide, and generate a two-dimensional or three-dimensional structure file of the linear peptide. This tool is generally used for creating small peptide structures.\n",
      "\n",
      "## 147. Peptide VS\n",
      "\n",
      "Peptide VS module integrates AutoDock Vina and AutoDock CrankPep for protein-polypeptide docking, thereby predicting the conformation of protein-polypeptide, obtaining the energy of molecular docking and binding affinity.\n",
      "\n",
      "## 148. Pocket Finder\n",
      "\n",
      "The Pocket Finder module identifies these pockets based on geometric and physicochemical properties, and its main function is to quickly and accurately identify potential pockets on the protein surface. Protein pockets (or active sites) are small areas on the surface of proteins, usually where drug molecules or other small molecules bind. Identifying these pockets is crucial for drug design and protein function research.\n",
      "\n",
      "## 149. Property Filter\n",
      "\n",
      "Property Filter is very versatile and can select a subset of molecules based either on properties imported with the molecule (as from a SDF file) or from calculations on the molecule on the fly. The supported input file formats are: SD (.sdf, .sd). The supported output file formats are: SD (.sdf, .sd).\n",
      "\n",
      "## 150. Protein BLAST\n",
      "\n",
      "Protein BLAST is a protein Blast database that integrates sequences from various databases including GenPept, Swissprot, PIR, PDF, PDB, and RefSeq.\n",
      "\n",
      "## 151. Protein Binder Design (RFDiffusion)\n",
      "\n",
      "Diffusion models have had considerable success in image and language generative modeling but limited success when applied to protein modeling, likely due to the complexity of protein backbone geometry and sequence-structure relationships. David Baker's group developed RoseTTAFold Diffusion (RFdiffusion), and demonstrate the power and generality of the method, by experimentally characterizing the structures and functions of hundreds of new designs. By fine-tuning the RoseTTAFold structure prediction network on protein structure denoising tasks, they obtain a generative model of protein backbones that achieves outstanding performance on unconditional and topology-constrained protein monomer design, protein binder design, symmetric oligomer design, enzyme active site scaffolding, and symmetric motif scaffolding for therapeutic and metal-binding protein design. In a manner analogous to networks which produce images from user-specified inputs, RFdiffusion enables the design of diverse, complex, functional proteins from simple molecular specifications.\n",
      "\n",
      "## 152. Protein Design (ABACUS-R)\n",
      "\n",
      "ABACUS-R is a method based on deep learning for designing amino acid sequences that autonomously fold into a given target backbone. Please cite if you find this module helpful: Liu, Y., Zhang, L., Wang, W. et al. Rotamer-free protein sequence design based on deep learning and self-consistency. Nat Comput Sci. 2022 June 21; 2:451–462.\n",
      "\n",
      "## 153. Protein Design (LigandMPNN)\n",
      "\n",
      "LigandMPNN is a deep learning-based protein sequence design method specifically designed to simulate interactions between proteins and non-protein components (such as small molecules, nucleotides, and metals). It is an upgraded version of ProteinMPNN and can incorporate non-protein components into protein design, thereby enhancing the understanding of non-protein-protein interactions.\n",
      "\n",
      "## 154. Protein Design (ProteinMPNN)\n",
      "\n",
      "ProteinMPNN is a deep learning-based protein sequence design method that achieves a sequence recovery rate of 52.4% on natural protein scaffolds, compared to 32.9% for Rosetta. Adding noise during the training process can improve the sequence recovery rate of the protein structural model, and the resulting sequences can more robustly encode their structures. X-ray crystallography, cryo-electron microscopy, and functional studies have also demonstrated the wide applicability and high accuracy of ProteinMPNN, which has successfully rescued previously failed protein monomers, cyclic homooligomers, tetrahedral nanoparticles, and target-binding proteins designed using Rosetta or AlphaFold.\n",
      "\n",
      "## 155. Protein Design (RFDesign)\n",
      "\n",
      "Protein Design (RFDesign) is a fairly general method for protein design (fantasy and modification) based on RoseTTAFold. In addition to providing the structure and sequence of functional sites that need to be designed, no other input is required. Unlike current non-deep learning methods, it does not require specification of the topology of secondary structure or backbone, and can generate both sequence and structure simultaneously.\n",
      "\n",
      "## 156. Protein Design (RFDiffusion)\n",
      "\n",
      "Diffusion models have had considerable success in image and language generative modeling but limited success when applied to protein modeling, likely due to the complexity of protein backbone geometry and sequence-structure relationships. David Baker's group developed RoseTTAFold Diffusion (RFdiffusion), and demonstrate the power and generality of the method, by experimentally characterizing the structures and functions of hundreds of new designs. By fine-tuning the RoseTTAFold structure prediction network on protein structure denoising tasks, they obtain a generative model of protein backbones that achieves outstanding performance on unconditional and topology-constrained protein monomer design, protein binder design, symmetric oligomer design, enzyme active site scaffolding, and symmetric motif scaffolding for therapeutic and metal-binding protein design. In a manner analogous to networks which produce images from user-specified inputs, RFdiffusion enables the design of diverse, complex, functional proteins from simple molecular specifications.\n",
      "\n",
      "## 157. Protein Docking (FRODOCK)\n",
      "\n",
      "FRODOCK is capable of performing detailed 6D docking between two protein structures. This approximation allows for the efficient generation of potential predictions regarding how these two proteins may interact with each other. Since it is a rigid-body approximation, it is only effective when the expected conformational changes upon binding are minimal and is typically used as the initial step for docking, followed by further flexible docking.\n",
      "\n",
      "## 158. Protein Docking (HDOCK)\n",
      "\n",
      "HDOCK is a fast protein-protein docking program that integrates homology search, template-based modeling, structure prediction, large molecule docking, and bioinformatics integration. The server employs a hybrid docking strategy to predict binding complexes between two types of molecules, such as proteins and nucleic acids. It utilizes a combination of template-based modeling and de novo docking algorithms for protein-protein and protein-DNA/RNA docking.\n",
      "\n",
      "## 159. Protein Isoelectric Point (pI)\n",
      "\n",
      "Protein Isoelectric Point module is used to calculate the isoelectric point of protein, that is, the pH at which a particular molecule carries no net electrical charge, is an critical parameter for many analytical biochemistry and proteomics techniques, especially for 2D gel electrophoresis (2D-PAGE), capillary isoelectric focusing (cIEF), X-ray crystallography and liquid chromatography–mass spectrometry (LC-MS)\n",
      "\n",
      "## 160. Protein Physico-chemical Properties\n",
      "\n",
      "Protein Physico-chemical Properties is a module for calculating the physicochemical properties of protein sequences. The computed properties include molecular weight, isoelectric point, extinction coefficient, instability index, aromaticity, grand average of hydropathicity (GRAVY), and secondary structure composition.\n",
      "\n",
      "## 161. Protein Protonation\n",
      "\n",
      "The protein Protonation module predicts the pKa value for each protein residue using PROPKA3 and determines the protonation state based on the pH values.\n",
      "\n",
      "## 162. Protein Sequence Generation (ProGen)\n",
      "\n",
      "ProGen, a language model that can generate protein sequences with a predictable function across large protein families, akin to generating grammatically and semantically correct natural language sentences on diverse topics. The model was trained on 280 million protein sequences from >19,000 families and is augmented with control tags specifying protein properties. ProGen can be further fine-tuned to curated sequences and tags to improve controllable generation performance of proteins from families with sufficient homologous samples.\n",
      "\n",
      "## 163. Protein Structure Prediction (AlphaFold2)\n",
      "\n",
      "AlphaFold2 is a highly accurate protein structure prediction tool and is one of the most precise methods currently available, even approaching experimental levels. Version 2.1.0.\n",
      "\n",
      "## 164. Protein Structure Prediction (AlphaFold2.3.0)\n",
      "\n",
      "AlphaFold2 is currently one of the best protein structure prediction methods in the industry developed by DeepMind. In the 2020 protein structure prediction competition CASP14, AlphaFold2 achieved a score of nearly 90 points, ranking first and significantly ahead of the second place. It predicts the structures of most proteins with a difference of only one atom from the true structure, reaching the level of human observation and prediction using complex instruments such as cryo-electron microscopy. This is an unprecedented breakthrough in the history of protein structure prediction. Subsequent updated versions support the prediction of complex structures, including protein-peptide complexes. AlphaFold2 is a highly accurate protein structure prediction package. Version: 2.3.0\n",
      "\n",
      "## 165. Protein Structure Prediction (AlphaFold2.3.2)\n",
      "\n",
      "AlphaFold2 is a highly accurate protein structure prediction package. This is a completely new model that was entered in CASP14 and published in Nature. Version: v2.3.2. It is recommended to use AF3-like modules (such as Boltz-1, Chai-1, HelixFold3, and Protenix).\n",
      "\n",
      "## 166. Protein Structure Prediction (ESMFold)\n",
      "\n",
      "ESMFold is a protein structure prediction model developed by Meta company, which uses a large language model to directly infer structure from the primary sequence. It predicts structures 60 times faster than AlphaFold while maintaining resolution and accuracy.\n",
      "\n",
      "## 167. Protein Structure Prediction (RaptorX-Single)\n",
      "\n",
      "Protein Structure Prediction (RaptorX-Single) module is implemented based on the RaptorX-Single algorithm, which is a single sequence-based protein structure prediction method that does not require multiple sequence alignment (MSA) information. It integrates multiple protein language models and a structure generation module. The results show that RaptorX-Single, in addition to running faster than MSA-based methods such as AlphaFold2, also outperforms AlphaFold2 and other MSA-free methods in predicting antibody structures, proteins with very few homologous sequences, and single mutation effects. RaptorX-Single also outperforms AlphaFold2 when predicting protein sequences with a large number of homologous sequences.\n",
      "\n",
      "## 168. Protein-Protein Interaction (STRING)\n",
      "\n",
      "Protein-Protein Interaction (STRING) is a module for extracting protein interactions based on STRING. STRING is a protein interaction network database that includes both direct physical interactions and indirect functional associations between proteins.\n",
      "\n",
      "## 169. RNA 3D Structure Prediction\n",
      "\n",
      "Build three-dimensional de novo models of RNAs of sizes up to ~300 nts, given secondary structure and experimental constraints. Besides the RNA sequence to predict, you also need to provide a secondary structure file: a text file with secondary structure described in the dot-parentheses notation.\n",
      "\n",
      "## 170. RNA Secondary Structure Prediction\n",
      "\n",
      "Predict secondary structures of single-stranded RNA or DNA sequences using dynamic programming algorithms which yield a single optimal structure and the minimum free energy.\n",
      "\n",
      "## 171. Ramachandran Plots\n",
      "\n",
      "Evaluate the quality of models after homology modeling, focusing on the reasonableness of the protein's conformation without considering energy issues.\n",
      "\n",
      "## 172. Receptor-Based Peptide Design\n",
      "\n",
      "Receptor-Based Peptide Design is used for binding peptide design based on receptor structure (currently single-strand supported).\n",
      "\n",
      "## 173. Replace Chain Name\n",
      "\n",
      "Performs in-place replacement of a chain identifier by another.\n",
      "\n",
      "## 174. Restrained Complex Structure Prediction\n",
      "\n",
      "The Restrained Complex Structure Prediction module is implemented based on the ColabDock framework, which significantly improves the accuracy of protein-protein docking prediction by integrating multiple experimental constraints.\n",
      "\n",
      "## 175. Retrosynthetic Prediction (AiZynthFinder)\n",
      "\n",
      "Retrosynthetic Prediction (AiZynthFinder) is a tool for retrosynthetic planning. The algorithm is based on a Monte Carlo tree search that recursively breaks down a molecule to purchasable precursors. The tree search is guided by a policy that suggests possible precursors by utilizing a neural network trained on a library of known reaction templates.\n",
      "\n",
      "## 176. SD File\n",
      "\n",
      "SD File is a module for specifying small molecule structure in SDF format which could be used for multiple modules.\n",
      "\n",
      "## 177. SDF File Split\n",
      "\n",
      "SDF File Split is a tool for splitting an SD File into multiple SD files. Each new SD File contains a compound subset of similar size from the initial file.\n",
      "\n",
      "## 178. SDF Viewer\n",
      "\n",
      "SDF Viewer is a visualization tool for the small molecular library. Generate an interactive HTML table with columns corresponding to molecules and available alphanumerical data in an input file.\n",
      "\n",
      "## 179. Salts Removal\n",
      "\n",
      "Remove salts from molecules or simply count the number of molecules containing salts.\n",
      "\n",
      "## 180. Scaffold Constrained Generation\n",
      "\n",
      "During the optimization of a lead series, it is common to have scaffold constraints imposed on the structure of the molecules designed. Without enforcing such constraints, the probability of generating molecules with the required scaffold is extremely low and hinders the practicality of generative models for de novo drug design.\n",
      "\n",
      "## 181. SeqKit\n",
      "\n",
      "Ultrafast comprehensive toolkit for FASTA/Q processing, rapidly accomplishing common FASTA/Q file manipulations.\n",
      "\n",
      "## 182. Sequence Clustering\n",
      "\n",
      "Sequence clustering uses the DBSCAN algorithm to perform cluster analysis on the results of multiple sequence alignment (MSA), dividing multiple sequences into multiple cluster categories, and using the visualization module UMAP to embed sequences and obtain two-dimensional visualization information.\n",
      "\n",
      "## 183. Sequence Embedding Generation\n",
      "\n",
      "The Sequence Embedding Generation module is based on ESMFold's large-scale pre-trained protein language model. embeddings of sequences can be extracted to predict downstream sequence properties (such as affinity changes and stability changes corresponding to mutations, developability of antibody sequences, etc.) and provide sequence features for the training of discriminant models. ESM model is a universal protein language model. Sequence databases such as UniRef50/90 (tens of millions of sequences) are used for model training, and various models with different reference numbers (8 million, 35 million, 150 million, 650 million, 3 billion, 15 billion) are provided. It can be used to predict structure, function, and other protein properties directly from protein sequences. For example, in structure prediction, ESM avoids the need for external evolutionary databases, MSA and templates, and the calculation accuracy is close to AlphaFold2 (when MSA information exists). When MSA information is not available, ESMFold is significantly better than AlphaFold2. Computations are tens of times faster than AlphaFold2.\n",
      "\n",
      "## 184. Sequence Mutagenesis (Directed)\n",
      "\n",
      "Generate mutants based on a template sequence in batch. User could define all mutation locations and the mutated amino acids in a text file.\n",
      "\n",
      "## 185. Sequence Mutagenesis (Directed) for Ab\n",
      "\n",
      "Generate sequences of mutated antibody sequences based on a template antibody sequence and a mutation file (json) listing all mutations (normally the mutation file is automatically generated by BLAST and MSA). This is useful for high-throughput antibody engineering design.\n",
      "\n",
      "## 186. Sequence Mutagenesis (Saturated)\n",
      "\n",
      "Enumerate all possible point mutations at specified positions in a protein sequence, and generate text files for all corresponding mutations and mutant sequence files.\n",
      "\n",
      "## 187. Sequence Mutation\n",
      "\n",
      "Sequence Mutation is a protein sequence mutation module used to generate mutated sequences in bulk for specific sites. Mutation strategies include position-based mutations, homology-based mutations, mutations based on antibody CDR regions, and mutations based on both antibody CDR regions and homology. The types of mutations supported include alanine scanning, histidine mutation, and saturation mutagenesis.\n",
      "\n",
      "## 188. Sequence Translation\n",
      "\n",
      "Sequence Translation is a tool for Translating DNA sequences into RNA and protein sequences.\n",
      "\n",
      "## 189. Small Molecule Minimization\n",
      "\n",
      "Small Molecule Minimization is a small molecule energy minimization optimization tool that generates optimized 3D structure. UFF or MMFF molecular forcefields could be used for energy minimization. Conformation sampling methods, SDG, ETDG, KDG, and ETKDG could be used for generating initial 3D coordinates.\n",
      "\n",
      "## 190. Solubility Score\n",
      "\n",
      "Sequence-based protein solubility prediction.\n",
      "\n",
      "## 191. Solvent Exposure (SASA)\n",
      "\n",
      "The Residue SASA Calculation module calculates the solvent accessible surface area of residue based on structure PDB file.\n",
      "\n",
      "## 192. Structural Alignment\n",
      "\n",
      "Structural Alignment is a tool for the sequence-based structural alignment of two proteins. Performs a global pairwise alignment between two sequences using the BLOSUM62 matrix and the Needleman-Wunsch algorithm. Returns the alignment, the sequence identity, and the residue mapping between both original sequences.\n",
      "\n",
      "## 193. Structural Alignment (USalign)\n",
      "\n",
      "Structural Alignment (USalign) is a structural superposition tool based on USalign.\n",
      "\n",
      "## 194. Structural Energy\n",
      "\n",
      "Calculate the energy of multiple protein structures based on a physical model (molecular mechanics empirical force field) and compare it with the reference protein.\n",
      "\n",
      "## 195. Structure Clustering\n",
      "\n",
      "Structure Clustering is a small molecule clustering molecules based on a variety of 2D fingerprints using hierarchical clustering methodology.\n",
      "\n",
      "## 196. Structure Evolution\n",
      "\n",
      "Structure Evolution module is implemented based on the ESMIF model. The ESMIF inverse folding model aims to predict protein sequences based on the atomic coordinates of the protein backbone. The model is trained on 12 million protein structures predicted by AlphaFold2, and it includes an invariant geometric input processing layer followed by a sequence-to-sequence Transformer. It achieves a 51% local sequence recovery rate for backbone sequences that remain invariant in structure and a 72% recovery rate for buried residues. The model is also trained with span masking, allowing it to tolerate missing backbone coordinates, thus enabling the prediction of sequences for partially masked structures. This module can be used for both affinity maturation and stability optimization.\n",
      "\n",
      "## 197. Structure Minimization\n",
      "\n",
      "Structure Minimization is a structure optimization module, supporting three methods: hydrogen atom optimization, amino acid side chain optimization, and overall optimization.\n",
      "\n",
      "## 198. Structure Mutagenesis\n",
      "\n",
      "The Structure Mutagenesis module obtains protein sequence information from protein structure files and performs saturation mutagenesis or alanine mutagenesis at specified sites to generate a mutation file and a mutated sequence file containing mutation information. This module is used for subsequent structural mutation analysis in other modules.\n",
      "\n",
      "## 199. Structure Prediction (Boltz-1)\n",
      "\n",
      "Developed based on the Boltz-1 model, Boltz-1 is an open-source deep learning model that integrates innovations in model architecture, speed optimization, and data processing. It achieves AlphaFold3-level accuracy in predicting the 3D structures of biomolecular complexes. Boltz-1 demonstrates performance comparable to state-of-the-art commercial models across a range of benchmarks, setting a new standard for commercially usable tools in structural biology.\n",
      "\n",
      "## 200. Structure Prediction (Chai-1)\n",
      "\n",
      "Structure prediction using Chai-1, supporting protein, dna, rna, ions, ligands.\n",
      "\n",
      "## 201. Structure Prediction (HelixFold3)\n",
      "\n",
      "HelixFold3, developed by the Baidu PaddleHelix team, is comparable to AlphaFold3 in the accuracy of structure prediction of conventional small molecule ligands, nucleic acid molecules (including DNA and RNA), and proteins.\n",
      "\n",
      "## 202. Structure Prediction (Protenix)\n",
      "\n",
      "Protenix is the PyTorch version of the AlphaFold3 model reproduced by the AML AI4Science team at ByteDance. Here is a summary of the main contributions from the ByteDance AML AI4Science team: - Model Performance: Protenix has been benchmarked against existing models, demonstrating strong performance in structure prediction across different types of molecules. As a fully open-source model, it enables researchers to generate new predictions and fine-tune the model to meet specific application needs. - Methodology: During the reproduction process, Protenix was implemented based on the description of AF3, optimizing some ambiguous steps, correcting typographical errors, and making targeted adjustments based on model behavior. By sharing our reproduction experience, we hope to support the community in further advancing the field based on these improvements. - Accessibility: Protenix has been open-sourced, providing model weights, inference code, and training code for research purposes.\n",
      "\n",
      "## 203. Structure Prediction (RFAA)\n",
      "\n",
      "The Biomolecular Structure Prediction (RFAA) module is implemented based on the RoseTTAFold All-Atom (RFAA) method. RFAA relies on the network framework of RoseTTAFold itself, with relevant information of small molecules being appropriately input. RFAA is comparable to AlphaFold2 in terms of accuracy in protein structure prediction and also performs exceptionally well in flexible backbone small molecule docking. It can provide reasonable predictions for protein structures, multiple nucleic acid chains, and small molecule complexes.\n",
      "\n",
      "## 204. Structure Prediction (WFold)\n",
      "\n",
      "A third-party implementation of an AlphaFold3-like structure prediction model, with prediction results generally consistent with those of AF3, and accuracy surpassing that of AF2.\n",
      "\n",
      "## 205. Structure Preparation\n",
      "\n",
      "Structure Preparation is a protein structure preparation module used for adding missing atoms and residues, as well as for protonation determination and hydrogenation of protein amino acid residues.\n",
      "\n",
      "## 206. Structure Relax\n",
      "\n",
      "The Structure Relax module is used to eliminate unreasonable conformations in crystal structures, optimize structures designed with point mutations, and compare energy levels of multiple structures. The module searches for the optimal conformation within the local energy barriers of a given 3D structure through multiple iterations of amino acid side chain rearrangement and energy minimization calculations.\n",
      "\n",
      "## 207. Substructure Search\n",
      "\n",
      "Substructure Search is a tool for structure searching against a small molecule library file using a specified substructure and writing out the matched molecules to an output file. Substructure search is a common operation in cheminformatics research and can also be used for virtual screening to search for molecules containing specific functional fragments from small molecule commercial libraries for subsequent experimental validation.\n",
      "\n",
      "## 208. Synthetic Accessibility Score\n",
      "\n",
      "The SA Score (synthetic accessibility score) is an index for evaluating the feasibility of compound synthesis, which indicates whether a compound is easy to synthesize. The synthesis difficulty of small molecules was evaluated with values ranging from 1 to 10. The closer to 1, the easier to synthesize, and the closer to 10, the more difficult to synthesize. SA Score has become a commonly used indicator to predict the synthetic feasibility of new compounds and accelerate compound screening and drug discovery processes.\n",
      "\n",
      "## 209. TCM Chemical Ingredients\n",
      "\n",
      "TCM Chemical Ingredients is a module for extracting chemical structures of Chinese herbs.\n",
      "\n",
      "## 210. TCR-pMHC Complex Structure Prediction\n",
      "\n",
      "The cellular immune system is a crucial component of the human immune response, utilizing T cell receptors (TCRs) to recognize peptide-form antigens presented by major histocompatibility complex (MHC) proteins. Accurately defining the structural basis of TCRs and their binding to peptide-MHC complexes can provide important insights into both normal and abnormal immune responses and assist in guiding the design of vaccines and immunotherapies. Given the limited number of experimentally determined TCR-peptide-MHC structures and the vast number of TCRs and antigen targets within each individual, accurate modeling methods are needed. This module is based on TCRmodel2, which optimizes TCR-peptide-MHC complex modeling on the foundation of AlphaFold. It achieves comparable or higher accuracy than native AlphaFold and other benchmark-based TCR-peptide-MHC modeling methods, completing complex structure predictions within 30 minutes.\n",
      "\n",
      "## 211. Target Prediction (FastTargetPred)\n",
      "\n",
      "Target Prediction (FastTargetPred) is a small molecule target prediction module based on 2D similarity. Active molecules and target data are derived from ChEMBL database.\n",
      "\n",
      "## 212. Target Prioritization (OpenTargets)\n",
      "\n",
      "Target Prioritization (OpenTargets) is a module for extracting disease-related target proteins, based on the OpenTarget database and its disease-target association scoring method.\n",
      "\n",
      "## 213. Target-based Cyclic Peptide Design\n",
      "\n",
      "The target-based Cyclic Peptide Design module is based on the Evobind2 model, which is based on AF2 and aims to design novel Head-to-Tail cyclic peptides based on Target protein sequences. The high affinity combination can be directly selected by the predicted confidence index, the binding site can be specified, and the adversarial design can be avoided by isomer evaluation, which greatly improves the success rate.\n",
      "\n",
      "## 214. Template-guided Structure Prediction\n",
      "\n",
      "Template-guided structure prediction is based on a custom protein structure Template, and colabfold is used to predict protein structure.\n",
      "\n",
      "## 215. Therapeutic Antibody Profiler\n",
      "\n",
      "The Therapeutic Antibody Profiler (TAP) compares your antibody variable domain sequence against multiple developability guidelines derived from clinical-stage therapeutic values. TAP calculates the following properties to see if your antibody design is commenserate with those of clinical-stage therapeutics: Patches of Surface Hydrophobicity (PSH) metric across the CDR Vicinity Patches of Positive Charge (PPC) metric across the CDR Vicinity Patches of Negative Charge (PNC) metric across the CDR Vicinity Structural Fv Charge Symmetry Parameter (SFvCSP)\n",
      "\n",
      "## 216. Thermostability Prediction\n",
      "\n",
      "Thermostability Prediction, developed based on TemBERTure, is a deep learning tool designed to predict protein thermostability, focusing on amino acid sequence analysis. It includes two models: TemBERTureCLS and TemBERTureTm. TemBERTureCLS is a classification model used to predict the thermal category of a protein sequence, determining whether it is thermophilic or non-thermophilic. TemBERTureTm is a regression model used to predict the melting temperature (Tm) of a protein based on its sequence. Both models are based on the protBERT-BFD language model, which has been pre-trained on a large dataset of protein sequences. By using an adapter-based fine-tuning approach, TemBERTure can efficiently and robustly adapt to specific tasks without the need for extensive retraining.\n",
      "\n",
      "## 217. Toxic Fragment Identification\n",
      "\n",
      "AlphaTox is a toxicity prediction and toxicity fragment detection module for small molecules. Toxicity fragments were collected from the reported literatures.\n",
      "\n",
      "## 218. Tumor Gene Expression (TCGA)\n",
      "\n",
      "Plot gene expression in normal and tumor tissues, based on databases TCGA and GTEx. It retrieves the expression of specified genes in tumors and normal tissues, counts and maps the gene expression differences of tumor cells, tumor tissues, and normal tissues, etc., to help drug target selection and decision-making.\n",
      "\n",
      "## 219. Venn Diagram Plot\n",
      "\n",
      "Venn Diagram Plot is a module for creating Venn diagrams, commonly used to compare the overlapping areas of two sets and to extract the common elements. For example, it can be used in traditional Chinese medicine network pharmacology analysis to extract the intersection of predicted targets of traditional Chinese medicine components and disease-related targets.\n",
      "\n",
      "## 220. mRNA 5'UTRs optimization\n",
      "\n",
      "mRNA 5' UTRs optimization is implemented based on the Smart5UTR model. Smart5UTR is a novel deep generative model designed to create N1-methyl-pseudouridine (m1Ψ) 5' UTRs in mRNA sequences. Smart5UTR utilizes a multi-task autoencoder framework to effectively generate 5' UTR sequences by leveraging latent features learned from large datasets. The performance of mRNAs designed by Smart5UTR has been validated through both in vitro and in vivo experiments. This powerful tool simplifies the design of m1Ψ-5' UTRs and contributes to the development of more effective mRNA therapies.\n",
      "\n",
      "## 221. mRNA Optimization (AlphaRNA)\n",
      "\n",
      "mRNA Optimization (AlphaRNA) can optimize mRNA sequences for better codon usage bias and more stable secondary structures, to enhance its expression level, half-life, antibody titer, etc. When inputting UTR, if fixed UTR is selected, only the CDS region will be optimized. If fixed UTR is not selected, both UTR and CDS will still be optimized simultaneously. If you do not want CDS to be paired with UTR and only optimize CDS itself, you do not need to input UTR, just input CDS.\n",
      "\n",
      "## 222. siRNA Designer\n",
      "\n",
      "siRNA Designer designs siRNA molecular sequences based on target gene sequences.\n",
      "\n",
      "# WeMol模块基本信息统计\n",
      "\n",
      "## 1. 2D Similarity Search\n",
      "\n",
      "2D Similarity Search模块是基于分子指纹进行二维相似度搜索的工具。根据不同指纹类型（Maccs Key、pharmacophore ﬁngerprints、extended connectivity fingerprints）计算得到的指纹向量或者向量字符串进行相似性搜索，从分子数据库中筛选出与模板分子相似（不相似）的化合物。\n",
      "\n",
      "## 2. 3-letter AA Conversion\n",
      "\n",
      "把三字母表示的氨基酸转换为单字母表示。\"\"ASP ILE VAL ASN\"\"转换为 \"\"DIVQ\"\".\n",
      "\n",
      "## 3. 3D Conf Generation (AlphaConf)\n",
      "\n",
      "3D Conf Generation (AlphaConf)是小分子三维构象搜索模块。三维构象搜索与生成技术主要用于对蛋白质结构域或者化合物结构进行高效的搜索，以用于结构设计或筛选。唯信通过采用一种全新的限制性结构片段定义方式进行分子三维构象的生成，精度优于同类算法。通过采用非重复构象生成方法，节省大量计算时间，计算速度远超同类算法。独特高效的构象压缩技术，较同类算法的存储空间降低400~800倍，适用于超大规模三维构象库的构建和超高通量虚拟筛选。\n",
      "\n",
      "## 4. AA Probability Prediction\n",
      "\n",
      "基于预训练的大规模蛋白质语言模型（也叫做PLM或pLLM），预测序列中每个氨基酸（AA）位置处20种AA出现的概率。与进化上更保守的AA类似，语言模型预测的高概率AA，有利于提升结构的稳定性、改善蛋白的折叠、提升蛋白质的表达能力等、甚至提升亲和力，比随机盲目突变具有潜在的优势。 - ESM为基于序列的PLM，适用于蛋白包括抗体； - IgLM为基于序列的PLM，只适用于抗体，可以指定种属（比如人）； - All in One同时使用ESM与IgLM进行计算； - ESMIF为结构感知的PLM，适用于蛋白包括抗体； - AntiFold为基于ESMIF使用抗体数据微调的模型，更适用于抗体或纳米抗体。 没有结构的时候，可以使用ESM、IgLM等纯序列模型；有结构或者预测了结构，可以使用结构感知的模型，在稳定性、亲和力等跟局部结构相关性更强的任务上表现更好。\n",
      "\n",
      "## 5. AC2SDF\n",
      "\n",
      "AC2SDF模块用于将AlphaConf模块生成的构象压缩二进制文件AC.GZ转为便于查看的SDF文件。\n",
      "\n",
      "## 6. ADMET Prediction\n",
      "\n",
      "基于通信消息传递神经网络的ADMET预测。CMPNN通过通信内核加强节点和边之间的消息交互来改进分子嵌入。此外，通过引入新的消息增强模块，丰富了消息生成过程。\n",
      "\n",
      "## 7. ADMET Prediction (v2)\n",
      "\n",
      "ADMET Prediction (v2)是一个基于机器学习的小分子ADMET性质预测模块。能快速批量预测小分子的ADMET性质，支持图注意力神经网络模型、轻量梯度提升树模型、随机森林模型、梯度提升树模型4种常见高效的机器学习算法，分子特征支持分子指纹以及分子描述符两种方法，能对小分子化合物库进行快速批量预测。模块支持27种ADMET性质，其中7种回归模型，20中分类模型。用户可以根据介绍文档中的预测性能数据，选择理想的机器学习算法和分子特征化方法。\n",
      "\n",
      "## 8. Absolute Folding Stability\n",
      "\n",
      "Absolute Folding Stability Prediction模块通过蛋白序列生成模型ESM-IF，预测蛋白质的绝对稳定性ΔG。模型在收集的数据集上进行测试，发现预测误差RMSE ≈ 1.5 kcal/mol，PR相关系数为0.7。这一进展降低了通过实验方法（如CD、荧光定量PCR等）获取蛋白质稳定性ΔG的成本，为计算预测蛋白质的折叠稳定性ΔG带来重大突破。\n",
      "\n",
      "## 9. Aggregation Score\n",
      "\n",
      "Aggregation Score用于预测蛋白质结构中的聚集倾向和蛋白质溶解度，通过考虑序列和结构来预测蛋白质中易聚集的位点，这对于球状蛋白质特别有用，其中容易聚集的位点可能埋藏在天然结构内并且序列不连续。通过考虑天然氨基酸的实验聚集倾向尺度，该方法可以准确预测蛋白质聚集倾向。\n",
      "\n",
      "## 10. Alanine Scan\n",
      "\n",
      "丙氨酸扫描可以将蛋白质的每一个残基分别突变为Ala，并计算丙氨酸突变导致的自由能变化。 它对于快速扫描很有用，因为残基中的极性相互作用和位阻在突变为丙氨酸时都会被破坏。\n",
      "\n",
      "## 11. Alanine Scan (MMPBSA v2)\n",
      "\n",
      "Alanine Scan (MMPBSA)是计算丙氨酸突变后的结合自由能。\n",
      "\n",
      "## 12. Alanine Scan (MMPBSA)\n",
      "\n",
      "Alanine Scan (MMPBSA)是计算丙氨酸突变后的结合自由能。\n",
      "\n",
      "## 13. AlphaAutoMD (GMX2023 v2)\n",
      "\n",
      "AlphaAutoMD是一个完全自动化的基于gromacs的分子动力学模块，用户提交PDB文件，指定力场、模拟时长等关键参数即可快速提交MD作业，该模块内部整合了多个计算模块，大部分参数采用默认参数。\n",
      "\n",
      "## 14. AlphaAutoMD (GMX2023)\n",
      "\n",
      "AlphaAutoMD是一个完全自动化的基于gromacs的分子动力学模块，用户提交PDB文件，指定力场、模拟时长等关键参数即可快速提交MD作业，该模块内部整合了多个计算模块，大部分参数采用默认参数。\n",
      "\n",
      "## 15. AlphaShape\n",
      "\n",
      "AlphaShape是基于分子三维形状和药效团的虚拟筛选模块，算法在三维构象的基础上进行基于分子三维相似性的虚拟筛选。通过结合高斯函数与深度神经网络模型，计算精度领先同类型商业算法。\n",
      "\n",
      "## 16. Antibody (Off-) Target Prediction\n",
      "\n",
      "基于结构相似性原理从抗原-抗体数据库中（相似抗体可能具有相似靶点）预测抗体的潜在靶点（脱靶效应）。\n",
      "\n",
      "## 17. Antibody Design (MEAN)\n",
      "\n",
      "Antibody Design (MEAN)模块基于MEAN模型实现，该模型采用多通道等变图注意力网络，可用于设计CDR的一维序列和三维结构。\n",
      "\n",
      "## 18. Antibody Fv Humanness\n",
      "\n",
      "使用深度学习模型对抗体Fv序列进行人源评分。Humanness Score从0（非人源）到1（人源）。训练数据基于NGS测序的数十亿条抗体Fv序列。\n",
      "\n",
      "## 19. Antibody Fv Structure Prediction (IgFold)\n",
      "\n",
      "IgFold是一种基于深度学习的快速预测抗体Fv结构的方法。IgFold由一个预先训练的语言模型和直接预测骨架原子坐标的图网络组成，该语言模型训练了558M个天然抗体序列。 IgFold在显著更短的时间内（不到一分钟）预测出与其他方法（包括AlphaFold）相似或更好质量的结构。 注意：输入的抗体Fv区抗体序列名称中必须包含重链标识符：H，Heavy，.H；轻链标识符:L，Light，.L。 已知问题：部分预测结构会比输入序列缺失个别氨基酸，请留意！\n",
      "\n",
      "## 20. Antibody Mutation Analysis\n",
      "\n",
      "Antibody Mutation Analysis是基于DeepAb的源自网络置信度的新型突变评分指标，排名靠前的突变提高了结合亲和力。该模型可用于广泛的抗体预测和设计任务。\n",
      "\n",
      "## 21. Antibody Numbering\n",
      "\n",
      "Antibody Numbering是抗体编号模块，用于注释抗体可变区（Fv）或恒定区（包括 Fc）， 支持几乎所有主流的抗体编号规则，如可变区广泛使用的Kabat、Chothia 和 IMGT，以及恒定区主要使用的EU规则。\n",
      "\n",
      "## 22. Antibody Numbering v2\n",
      "\n",
      "Antibody Numbering v2是抗体编号模块，用于注释抗体可变区（Fv）或恒定区（包括 Fc）， 支持几乎所有主流的抗体编号规则，如可变区广泛使用的Kabat、Chothia 和 IMGT，以及恒定区主要使用的EU规则。\n",
      "\n",
      "## 23. Antibody Paratope Prediction\n",
      "\n",
      "预测抗体上与抗原结合的氨基酸位点（称为Paratope），基于等变图神经网络的深度学习模型，使用抗体结构进行训练和预测，预测精度在现有方法中最佳。\n",
      "\n",
      "## 24. Antibody RMSD\n",
      "\n",
      "Antibody RMSD模块对参考抗体结构及其他CDR相同的抗体结构，进行基于CDR区域的结构叠合，并计算CDR区域的RMSD值。\n",
      "\n",
      "## 25. Antibody Sequence Prediction (AbLang)\n",
      "\n",
      "根据OAS数据库中的抗体序列训练的语言模型，可预测抗体序列中指定位点可能的氨基酸，或者修复抗体序列数据中缺失残基，在抗体序列预测中优于通用蛋白质语言模型（如Meta开发的ESM-1b模型）。\n",
      "\n",
      "## 26. Antibody Sequence Prediction (IgLM)\n",
      "\n",
      "Antibody Sequence Prediction (IgLM)模块是抗体序列生成与优化，该方法从Observed Antibody Space (OAS) 收集抗体序列。OAS数据库包含六个物种的天然抗体序列：人类、小鼠、大鼠、兔子、恒河猴和骆驼。为了研究模型能力的影响，训练了两个版本的模型： IgLM和IgLM-S，分别有13M和1.4M的训练参数。两个IgLM模型都在558万非冗余序列上训练，这些序列基于95%相似性聚类。在训练过程中，随机屏蔽了抗体序列中10到20个残基，以便在推理过程中实现任意跨度的多样化。此外，还对序列中的链型（重链或轻链）和原产物种进行了限定，提供这样的背景能够控制物种特异性抗体序列的产生。该方法被证明可以从各种物种中产生全长的重链和轻链序列，以及具有改进可开发性的填充CDR环库。该方法是一个强大的抗体设计工具，可应用于各种抗体序列设计场景。\n",
      "\n",
      "## 27. Antibody Viscosity Prediction\n",
      "\n",
      "粘度是影响抗体药物开发的重要因素，临床上抗体往往需要静脉内或皮下给药，需要高浓度的抗体溶液（>100mg/mL）才能以小剂量注射获得与治疗相关的剂量，但是高浓度的抗体往往表现出高粘度，这对抗体药物的开发，制造和给药提出了挑战。研究发现，抗体序列是决定抗体粘度的关键因素，文献报道抗体粘度与Fv区域的电荷、VH和VL区域电荷的不对称性FvCSP和Fv区域的疏水指数HI存在相关性，基于抗体序列预测抗体粘度是一个有效方法。\n",
      "\n",
      "## 28. Antibody-Antigen Docking (HADDOCK)\n",
      "\n",
      "HADDOCK v3.0是对长期存在且经过时间验证的用于生物分子复合物集成建模的HADDOCK的自下而上的重新想象。这个年轻且仍处于实验阶段(使用它的风险由您自己承担!)的目标是模块化和扩展HADDOCK的核心功能。在目前的实现中，HADDOCK v3.0仍然缺乏生产web服务器版本HADDOCK v2.4所具有的全部功能。然而，它能够充分利用模糊交互约束(AIRs)来驱动对接过程。\n",
      "\n",
      "## 29. Back Mutation Grouping\n",
      "\n",
      "Back Mutation Grouping是抗体人源化设计流程中分组模块，根据Mutation Score模块输出的回复突变评分表对回复突变进行分组，并返回突变后的序列。\n",
      "\n",
      "## 30. Back Mutation Grouping (v2.3)\n",
      "\n",
      "Back Mutation Grouping是抗体人源化设计流程中分组模块，根据Mutation Score模块输出的回复突变评分表对回复突变进行分组，并返回突变后的序列。\n",
      "\n",
      "## 31. Back Mutation Grouping (v2.4)\n",
      "\n",
      "Back Mutation Grouping是抗体人源化设计流程中分组模块，根据Mutation Score模块输出的回复突变评分表对回复突变进行分组，并返回突变后的序列。\n",
      "\n",
      "## 32. Batch Renaming\n",
      "\n",
      "Batch Renaming模块设计用于化学库的分子重命名。用户可以使用前缀和定义的长度来规范分子名称。例如，将一个从WCP0001开始的库重命名为WCP9999，用户可以输入WCP前缀，长度为4。用户还可以使用——keeptitle参数保存以前的名称，以保存名称之间的关系。该模块可用于大型从头库或用户私有化学库中的自定义分子命名。\n",
      "\n",
      "## 33. CIF2PDB\n",
      "\n",
      "CIF2PDB模块是基于BioPython将mmCIF文件转换成PDB文件。\n",
      "\n",
      "## 34. Cleavage Site Prediction\n",
      "\n",
      "预测八种常用蛋白酶的蛋白型裂解位点，包括胰蛋白酶（trypsin），精氨酸C端肽段（ArgC），粒胰蛋白酶（chymotrypsin），谷氨酸C端蛋白酶（GluC），赖氨酸C端肽段（LysC），天冬氨酸N端肽段（AspN），赖氨酸N端肽段（LysN），L-精氨酸胺基肽酶（LysargiNase）。\n",
      "\n",
      "## 35. Codon Optimization\n",
      "\n",
      "Codon Optimization可用于密码子优化（基于PCR的基因合成的自动寡核苷酸设计）。整个基因组序列的可用性极大地增加了蛋白质靶标的数量，其中许多需要在原始DNA来源以外的细胞中过度表达。合成基因可以针对表达进行优化，并构建为易于突变操作而无需考虑亲本基因组。然而，合成基因的设计和构建，尤其是那些编码大蛋白质的基因，可能是一个缓慢、困难和令人困惑的过程。该模块通过基于PCR的方法自动设计用于基因合成的寡核苷酸。\n",
      "\n",
      "## 36. Complex Interface Analysis\n",
      "\n",
      "Complex Interface Analysis模块是基于结构的分析蛋白质复合物相互作用界面的关键残基。\n",
      "\n",
      "## 37. Cyclic Peptide Design\n",
      "\n",
      "Cyclic Peptide Design模块采用的是哈佛大学Sergey Ovchinnikov和华盛顿大学Gaurav Bhardwaj研究组开发的环肽设计算法AfCycDesign。AfCycDesign通过修改AlphaFold网络来准确预测和设计环肽。该模块可基于环肽模板分子结构的骨架进行环肽设计，也可以全新环肽设计。在包含49个环肽的数据集中，算法预测的36个结构具有高置信度。使用 AlphaFold 重新设计大环骨架序列，结晶测试显示算法设计的具有不同大小和结构的七个序列和 X 射线晶体结构非常匹配。\n",
      "\n",
      "## 38. Cyclic Peptide Structure Prediction\n",
      "\n",
      "Cyclic Peptide Structure Prediction模块利用线性肽的序列生成环肽的结构。\n",
      "\n",
      "## 39. De novo Generation\n",
      "\n",
      "REINVENT是一种用于分子全新设计的基于序列的生成模型调优方法，通过增强的情节似然性，可以学习生成具有特定指定理想性质的结构。我们展示了该模型如何执行一系列任务，例如生成类似于查询结构的类似物和生成预测对生物靶点具有活性的化合物。作为一个原理证明，该模型首先被训练生成不含硫的分子。作为第二个例子，该模型被训练生成与药物Celecoxib类似的分子，这种技术可以用于从单个分子开始进行支架跳跃或库扩展。最后，当将模型调整为生成预测对多巴胺受体类型2具有活性的化合物时，模型生成的结构中超过95%被预测为活性，包括在生成模型和活性预测模型中都未包含的实验确认的活性分子。\n",
      "\n",
      "## 40. De novo Generation (Moses)\n",
      "\n",
      "De novo Generation (Moses)是基于深度学习的分子生成模块，实现了多种主流的分子生成模型，包括字符级循环神经网络，变分自编码器，以及对抗自编码器。\n",
      "\n",
      "## 41. De novo Generation (REINVENT4)\n",
      "\n",
      "基于REINVENT4的小分子全新分子生成。支持多种分子生成方式：Reinvent - 从头开始创造新分子，Libinvent - 修饰一个骨架，Linkinvent - 设计两个片段之间的linker，Mol2Mol - 在用户定义的相似度范围内优化分子。\n",
      "\n",
      "## 42. De-immunization Design\n",
      "\n",
      "在具有免疫原性的序列上引入逐渐增加的突变（通常建议人源的突变），以降低或者去除免疫原性。 本模块通常不单独使用，一般结合AlphaMHC进行使用，自动化去除免疫原性请使用【De-immunization Design】流程。\n",
      "\n",
      "## 43. Descriptors (RDKit)\n",
      "\n",
      "Descriptors (RDKit)模块是计算分子的2D/3D描述符并将其写入SD或CSV/TSV文本文件中。2D描述符：Autocorr2D、MolWt、Ipc、NumRotatableBonds、qed等；3D描述符：Autocorr3D、RadiusOfGyration、Eccentricity等；以及FragmentCountOnly描述符：fr_Al_COO、fr_Al_OH、fr_Al_OH_noTert等。\n",
      "\n",
      "## 44. Disulfide Bond Search\n",
      "\n",
      "Disulfide Bond Search模块计算蛋白质中潜在的二硫键位置，这对优化蛋白质的稳定性有所作用。二硫键作为对蛋白质的稳定性有极大的作用，但是加入不合理的二硫键也会容易引起聚集，表达量降低甚至错误折叠等不利影响。\n",
      "\n",
      "## 45. Diverse Subset\n",
      "\n",
      "基于多种2D指纹以及使用最大最小距离（MaxMin）或分层聚类方法（Hierarchical Clustering）选择分子子集，并将它们写入文件。RDKit中可用的Dice和Tanimoto相似性函数能够处理对应于IntVect和BitVect的指纹。然而，所有其他相似性函数都期望使用BitVect指纹来计算成对相似性。因此，对于AtomPairs、Morgan、MorganFeatures和TopologicalTorsions的相似性计算，使用ExplicitBitVect指纹代替默认的IntVect指纹。\n",
      "\n",
      "## 46. Duplicates Removal\n",
      "\n",
      "Duplicates Removal模块是基于规范SMILES字符串识别和删除重复分子，或者仅统计重复分子数量。\n",
      "\n",
      "## 47. Electrostatic Potential Calculation (APBS)\n",
      "\n",
      "基于APBS方法计算生物大分子结构的静电势能，并绘制表面图。 为了可视化显示表面图，请从结构编辑器WeView中执行该功能：Weview->Analysis->Electrostatics。\n",
      "\n",
      "## 48. Enumerate Stereoisomers\n",
      "\n",
      "Enumerate Stereoisomers是枚举小分子立体异构体的工具，支持顺反异构体和对映异构体两种形式的枚举。\n",
      "\n",
      "## 49. Evaluate Nucleic Acid (AlphaRNA)\n",
      "\n",
      "Evaluate Nucleic Acid (AlphaRNA)模块用于评估核酸序列的其表达量和半衰期、抗体滴度等。支持人、小鼠、大鼠、猪等种属。\n",
      "\n",
      "## 50. Extract Fv Sequence\n",
      "\n",
      "Extract Fv Sequence是从抗体全长序列中提取Fv区序列的工具。\n",
      "\n",
      "## 51. Extract Sequence from Structure (PDB2FASTA)\n",
      "\n",
      "Extract Sequence from Structure (PDB2FASTA)模块是从蛋白的PDB文件中将序列提取出来保存为FASTA文件。常规氨基酸序列用单字母表示，其他类型都标注为X。\n",
      "\n",
      "## 52. FASTA File\n",
      "\n",
      "FASTA文件是一个用于指定fasta文件的模块，可用于其他模块的输入。会对FASTA文件的有效性进行判断。\n",
      "\n",
      "## 53. File\n",
      "\n",
      "File是用于指定输入文件的模块，可用于多个模块的统一输入。\n",
      "\n",
      "## 54. Format Conversion (Open Babel)\n",
      "\n",
      "Format Conversion (Open Babel)是基于Open Babel的格式转换模块，主要用于处理各种化学数据。允许任何人从分子建模、化学、固态材料、生物化学或相关领域搜索、转换、分析或存储数据。\n",
      "\n",
      "## 55. Format Conversion (RDKit)\n",
      "\n",
      "Format Conversion (RDKit)是基于RDKit的分子文件格式转换工具。\n",
      "\n",
      "## 56. Free Energy Surface Analysis\n",
      "\n",
      "基于PLUMED元动力学模拟后的自由能计算。\n",
      "\n",
      "## 57. GMX Ligand Parameterization\n",
      "\n",
      "GMX Ligand Parameterization模块根据小分子pdb文件生成分子动力学模拟（GROMACS）所需的MOL2，GRO以及ITP文件。\n",
      "\n",
      "## 58. GMX MD Run (GMX2023)\n",
      "\n",
      "GMX MD Run (GMX2023)模块是利用已经准备好的体系拓扑文件以及参数文件进行基于GROMACS的分子动力学模拟。\n",
      "\n",
      "## 59. GMX MDP Generation (Auto)\n",
      "\n",
      "GMX MDP Generation (Auto)模块主要是根据所选体系（膜，受体，配体）自动生成分子动力学模拟过程中所需的MDP文件，此文件是Gromacs分子动力学模拟需要用到输入文件，里面包含各种参数。若需要设置更细节的参数，请前往Minimize MDP Generation，NPT MDP Generation，MD MDP Generation模块。\n",
      "\n",
      "## 60. GMX Membrane Parameterization\n",
      "\n",
      "GMX Membrane Parameterization模块是根据Amber或者Charmm生成膜结构的GRO，ITP以及TOP文件。\n",
      "\n",
      "## 61. GMX Metadynamics Generation\n",
      "\n",
      "GMX Metadynamics Generation模块是生成可用于Metadynamics模拟的输入文件。\n",
      "\n",
      "## 62. GMX Receptor Parameterization\n",
      "\n",
      "根据Gromacs生成受体（包括蛋白或者核酸）的GRO，ITP以及TOP文件。\n",
      "\n",
      "## 63. Gene Enrichment (DAVID)\n",
      "\n",
      "Gene Enrichment (DAVID)是基于DAVID的基因功能富集分析模块，DAVID是一个生物信息数据库，整合了生物学数据和分析工具，为大规模的基因或蛋白列表提供系统综合的生物功能注释信息。\n",
      "\n",
      "## 64. Generate Humanized Variants\n",
      "\n",
      "抗体人源化设计中基于Grafting以及Back Mutation Grouping的结果批量生成人源化后的序列。\n",
      "\n",
      "## 65. Germline AA Distribution Frequency\n",
      "\n",
      "Germline AA Distribution Frequency模块输出抗体各位置的germline的氨基酸频率分布。可以按指定的germline基因家族分别输出（通常关注与目标序列同家族germline基因的频率分布情况）。\n",
      "\n",
      "## 66. Germline Blast\n",
      "\n",
      "Germline Blast模块基于IgBlastp实现，通过序列比对从IMGT reference sequences数据库中搜索与目标抗体序列最接近的同源模板，输出对应的模板序列以及序列一致性等信息。数据库中默认检索的序列类型为：IMGT V genes(F+ORF+in-frame P)。\n",
      "\n",
      "## 67. Grafting (v2)\n",
      "\n",
      "Grafting模块是移植抗体的CDR到特定的框架区模板上，通常用于人源化设计。\n",
      "\n",
      "## 68. Grafting (v2.1)\n",
      "\n",
      "Grafting模块是移植抗体的CDR到特定的框架区模板上，通常用于人源化设计。版本：v2.1\n",
      "\n",
      "## 69. Grafting (v2.2)\n",
      "\n",
      "Grafting模块是移植抗体的CDR到特定的框架区模板上，通常用于人源化设计。版本：v2.2\n",
      "\n",
      "## 70. Grafting (v2.3)\n",
      "\n",
      "Grafting模块是移植抗体的CDR到特定的框架区模板上，通常用于人源化设计。版本：v2.3\n",
      "\n",
      "## 71. Grafting (v2.4)\n",
      "\n",
      "Grafting模块是移植抗体的CDR到特定的框架区模板上，通常用于人源化设计。版本：v2.4\n",
      "\n",
      "## 72. Homology Modeling (Protein)\n",
      "\n",
      "蛋白质三维结构的同源性或比较建模。用户提供一个待建模的序列与已知相关结构的比对。通过满足空间约束条件进行比较蛋白质结构建模，以及许多其他任务，包括蛋白质结构中循环的全新建模、针对灵活定义的目标函数优化各种蛋白质结构模型、蛋白质序列和/或结构的多重比对、聚类、搜索序列数据库、比较蛋白质结构等。\n",
      "\n",
      "## 73. Homology Tree\n",
      "\n",
      "Homology Tree模块用于生成同源性进化树。\n",
      "\n",
      "## 74. Human Antibody BLAST\n",
      "\n",
      "Human Antibody BLAST是人类抗体数据库Blast模块，该数据库目前包含来自超过75项不同研究的超过10亿个序列，涵盖了来自人类的多种免疫状态和个体。提交抗体序列，将返回同源性最高的人源同源抗体序列，可用于高级抗体人源化设计、亲和力成熟、去免疫原性、抗体工程等。\n",
      "\n",
      "## 75. Human Fragment BLAST\n",
      "\n",
      "Human Fragment BLAST是基于输入的九肽, 在人源片段库(Germline, TCR, NextProt, OAS)中搜索最相似的9肽。\n",
      "\n",
      "## 76. Human Germline BLAST\n",
      "\n",
      "通过序列比对在人类生殖系数据库中搜索与目标抗体序列接近的同源模板，输出对应的模板序列以及序列一致性信息。\n",
      "\n",
      "## 77. Human Germline BLAST (v2.0)\n",
      "\n",
      "通过序列比对在人类生殖系数据库中搜索与目标抗体序列接近的同源模板，输出对应的模板序列以及序列一致性信息。\n",
      "\n",
      "## 78. Humanization Report\n",
      "\n",
      "Humanization Report是抗体人源化设计报告生成模块，用于生成最终的抗体人源化设计报告以及相应的专利实施例段落。\n",
      "\n",
      "## 79. Humanization Report (v2.3)\n",
      "\n",
      "抗体人源化设计报告生成模块，用于生成最终的抗体人源化设计报告以及相应的专利实施例段落。\n",
      "\n",
      "## 80. Humanization Report (v2.4)\n",
      "\n",
      "抗体人源化设计报告生成模块，用于生成最终的抗体人源化设计报告以及相应的专利实施例段落。相比v2.3，新增RMSD和能量信息。\n",
      "\n",
      "## 81. IgG Modeling\n",
      "\n",
      "IgG Modeling对抗体全长序列进行建模，用于构建抗体IgG完整的三维结构，支持单特异性和双特异性抗体。 自动识别全长序列中的可变区（Fv）序列并通过SOTA的方法（目前为ESMFold）进行建模，IgG的其余部分包括Fc和linker以已知全长抗体的晶体结构为模板通过空间约束条件进行同源模建，效果比直接用AF2等方法预测完整IgG结构更优。\n",
      "\n",
      "## 82. Immune Protein Structure Prediction\n",
      "\n",
      "Immune Protein Structure Prediction模块是基于ImmuneBuilder算法预测免疫蛋白的结构。ImmuneBuilder是一组深度学习模型，专门预测抗体、纳米抗体和T细胞受体的结构，精度高的同时比AlphaFold2快得多。\n",
      "\n",
      "## 83. Immunogenicity Prediction (AlphaMHC v2.0)\n",
      "\n",
      "AlphaMHC算法采用流行的NLP自然语言处理技术，全新的多模融合深度神经网络架构，整合了近10亿条公开及私有的与免疫原性相关的湿实验数据（包括亲和力数据、NGS数据、质谱数据等）进行训练，成功实现了从序列到临床免疫原性风险的端到端的预测，并通过上百条来自FDA、EMA的临床真实免疫原性数据（包括单/多特异性抗体和重组蛋白等）进行验证，AlphaMHC能够准确区分免疫原性的高低，ROC-AUC达0.87，准确性超过80%（部分测试集高达91%），表现出比现有方法显著更优的预测性能，是已知唯一一个可以得到高质量临床数据验证的算法。 注：推荐在WeSeq序列编辑器中调用此功能（Immunogenicity按钮），可以在序列中直观看到T细胞表位的位置。\n",
      "\n",
      "## 84. Immunogenicity Prediction (AlphaMHC v3.0 beta)\n",
      "\n",
      "AlphaMHC算法采用流行的NLP自然语言处理技术，全新的多模融合深度神经网络架构，整合了近10亿条与免疫原性相关的湿实验数据（包括亲和力数据、NGS数据、质谱数据等）进行训练，实现了从序列到临床免疫原性风险的端到端的预测，并通过上百条来自FDA、EMA的临床真实免疫原性数据（包括单/多特异性抗体和重组蛋白等）进行验证。 该版本使用抗体为主的临床ADA数据进行测试精度达到90%，AUC达0.91，性能优于v2.0版本。\n",
      "\n",
      "## 85. Immunogenicity Prediction (WeADApt v4.1.0)\n",
      "\n",
      "唯信自研最新免疫原性预测算法的内测版：AlphaMHC v4.1.0 beta。\n",
      "\n",
      "## 86. MD (GMX2024)\n",
      "\n",
      "GMX MD Run (GMX2024)模块是利用已经准备好的体系拓扑文件以及参数文件进行基于GROMACS的分子动力学模拟。\n",
      "\n",
      "## 87. MD Clustering\n",
      "\n",
      "MD Clustering是对动力学轨迹进行归簇分析。\n",
      "\n",
      "## 88. MD Clustering (v2)\n",
      "\n",
      "MD Clustering是对动力学轨迹进行归簇分析。\n",
      "\n",
      "## 89. MD Distance\n",
      "\n",
      "MD Distance是分子动力学轨迹的距离分析模块，输出分子动力学过程中两个组之间距离 (质心距离或几何中心距离) 随时间的变化。\n",
      "\n",
      "## 90. MD Distance (v2)\n",
      "\n",
      "MD Distance是分子动力学轨迹的距离分析模块，输出分子动力学过程中两个组之间距离 (质心距离或几何中心距离) 随时间的变化。\n",
      "\n",
      "## 91. MD Gyration\n",
      "\n",
      "MD Gyration回旋半径分析，可用来衡量体系模拟时的质权平均半径。\n",
      "\n",
      "## 92. MD Gyration (v2)\n",
      "\n",
      "MD Gyration回旋半径分析，可用来衡量体系模拟时的质权平均半径。\n",
      "\n",
      "## 93. MD Hbond\n",
      "\n",
      "MD Hbond对于指定组别之间的氢键分析。\n",
      "\n",
      "## 94. MD Hbond (v2)\n",
      "\n",
      "MD Hbond对于指定组别之间的氢键分析。\n",
      "\n",
      "## 95. MD MDP Generation\n",
      "\n",
      "MD MDP Generation模块主要是生成平衡模拟（MD）的MDP文件，此文件是Gromacs分子动力学模拟需要用到输入文件，里面包含各种参数。\n",
      "\n",
      "## 96. MD PCA\n",
      "\n",
      "MD PCA（Principal component analysis，PCA）模块可以从高维数据中分析出主要的影响因素 (本征向量) 。前几个本征向量（主成分，如前两个主成份则为 PC1，PC2) 一般可以描述分子运动的大部分信息。N个原子的柔性大体系如蛋白，其运动轨迹需要3N维笛卡尔坐标来描述，这样高维的数据很难理解和直观分析。\n",
      "\n",
      "## 97. MD PCA (v2)\n",
      "\n",
      "MD PCA（Principal component analysis，PCA）模块可以从高维数据中分析出主要的影响因素 (本征向量) 。前几个本征向量（主成分，如前两个主成份则为 PC1，PC2) 一般可以描述分子运动的大部分信息。N个原子的柔性大体系如蛋白，其运动轨迹需要3N维笛卡尔坐标来描述，这样高维的数据很难理解和直观分析。\n",
      "\n",
      "## 98. MD PDB Prepare\n",
      "\n",
      "MD PDB Prepare是一个在分子动力学模拟前PDB结构处理模块，结合PDBFixer工具对输入PDB文件中的蛋白结果进行修复，再分离出PDB文件中的蛋白结构、小分子结构以及核酸结构。\n",
      "\n",
      "## 99. MD RMS\n",
      "\n",
      "MD RMS模块是通过计算平衡模拟轨迹的均方根偏差（RMSD，Root Mean Square Deviation）和均方根波动（RMSF，Root Mean Square Fluctuation），从而分析结构的稳定性和结构变化情况。\n",
      "\n",
      "## 100. MD RMS (v2)\n",
      "\n",
      "MD RMS模块是通过计算平衡模拟轨迹的均方根偏差（RMSD，Root Mean Square Deviation）和均方根波动（RMSF，Root Mean Square Fluctuation），从而分析结构的稳定性和结构变化情况。\n",
      "\n",
      "## 101. MD SASA\n",
      "\n",
      "MD SASA模块是计算指定组别的溶剂可及表面积（solvent accessible surface area，SASA）。\n",
      "\n",
      "## 102. MD SASA (v2)\n",
      "\n",
      "MD SASA模块是计算指定组别的溶剂可及表面积（solvent accessible surface area，SASA）。\n",
      "\n",
      "## 103. MD Solvation\n",
      "\n",
      "MD Solvation对输入的受体配体文件加入水盒子和离子。\n",
      "\n",
      "## 104. MD Solvation v2\n",
      "\n",
      "对MD体系加入水盒子和离子。v2新增自主添加金属离子环境功能。\n",
      "\n",
      "## 105. MD Trajectory\n",
      "\n",
      "可根据起始帧数、结束帧数以及间隔帧数对平衡模拟进行轨迹提取，从而将其转换为GRO或者PDB轨迹文件。\n",
      "\n",
      "## 106. MD Trajectory (v2)\n",
      "\n",
      "可根据起始帧数、结束帧数以及间隔帧数对平衡模拟进行轨迹提取，从而将其转换为GRO或者PDB轨迹文件。\n",
      "\n",
      "## 107. MHC-I Binding Prediction\n",
      "\n",
      "MHC-I Binding Prediction是基于神经网络的预测MHC-I型亲和力预测模型。模型训练是利用亲和力(BA)和质谱洗脱配体（MS eluted ligand）的数据，基于NNAlign框架增加了预测特定MHC分子结合肽段的亲和力值和肽段的长度。NetMHCpan-4.0的方法提高了在肿瘤新抗原，验证的洗脱配体（ELs），T细胞免疫表位的预测准确性。\n",
      "\n",
      "## 108. MMPBSA\n",
      "\n",
      "MMPBSA计算受体与配体之间的结合自由能，并且提供能量分解数据、结合常数（Ka）、抑制剂常数（Ki）。\n",
      "\n",
      "## 109. MMPBSA (v2)\n",
      "\n",
      "MMPBSA计算受体与配体之间的结合自由能，并且提供能量分解数据、结合常数（Ka）、抑制剂常数（Ki）。\n",
      "\n",
      "## 110. Membrane Solvation\n",
      "\n",
      "Membrane Solvation对输入的膜，受体，配体文件加入水盒子和离子。\n",
      "\n",
      "## 111. Membrane System Construction\n",
      "\n",
      "Membrane System Construction构建膜结构的PDB文件。\n",
      "\n",
      "## 112. Metabolism Site Prediction\n",
      "\n",
      "Metabolism Site Prediction模块为预测小分子被P450代谢的代谢位点。模型对小分子的每个原子进行评估被代谢的可能性，并通过打分排序。支持的小分子输入文件格式为：SD（.sdf、.sd）、SMILES（.smi）。\n",
      "\n",
      "## 113. Minimize MDP Generation\n",
      "\n",
      "Minimize MDP Generation模块主要是生成Gromacs分子动力学模拟需要用到体系能量优化（Minimization）的输入MDP文件。\n",
      "\n",
      "## 114. Molecular Docking (AutoDock-GPU)\n",
      "\n",
      "Molecular Docking (AutoDock-GPU)是基于AutoDock的分子对接工具，采用GPU加速版本，主要用于预测分子之间的结合模式和相互作用，得到分子对接的能量和结合亲和力等信息。它还可以计算和比较多个分子之间的结合能力，用于药物分子的筛选、设计和优化。AutoDock-GPU是AutoDock4.2.6的OpenCL和Cuda加速版本，其利用可并行的LGA，从而通过在多个计算单元上并行处理配体-受体结合构象。配体文件支持的输入格式为SD（.sdf, .sd）、PDB（.pdb）和MOL（.mol）。受体结构文件支持的输入格式为PDB（.pdb）。\n",
      "\n",
      "## 115. Molecular Docking (DiffDock)\n",
      "\n",
      "Molecular Docking (DiffDock)是一种扩散生成模型，主要用于小分子和蛋白对接。DiffDock在PDBBind上获得了38%的top-1成功率（RMSD<2A），大大超过了以前传统对接（23%）和深度学习（20%）方法的最先进水平。此外，以前的方法无法对接计算上的折叠结构（最大精度为10.4%），而DiffDock保持了明显更高的精度（21.7%）。最后，DiffDock具有快速的推理时间，并提供具有高选择性精度的置信度估计值。\n",
      "\n",
      "## 116. Molecule In Membrane\n",
      "\n",
      "Molecule In Membrane模块是生成受体/配体/复合物取向位置的结构文件。\n",
      "\n",
      "## 117. Multiple Sequence Alignment\n",
      "\n",
      "基于渐进（progressive）比对算法进行多重序列比对，绘制进化树与序列对比图。\n",
      "\n",
      "## 118. Multiple Sequence Alignment (MAFFT)\n",
      "\n",
      "基于MAFFT的多序列比对程序，支持蛋白和核酸序列的比对。\n",
      "\n",
      "## 119. Mutation Energy of Binding\n",
      "\n",
      "Mutation Energy of Binding模块旨在计算突变对复合物结合能的影响。根据输入的复合物结构及突变文件构建突变结构，并计算链之间的结合能，与野生型对比，计算突变前后链之间的结合能变化。能量越负，说明突变越有利于指定链之间的结合。\n",
      "\n",
      "## 120. Mutation Energy of Binding (GeoPPI)\n",
      "\n",
      "Mutation Energy of Binding (GeoPPI)模块是基于深度学习的框架，使用蛋白质复合物的深度几何表征来模拟突变对结合亲和力的影响，从而预测氨基酸突变对蛋白质-蛋白质亲和力的影响。\n",
      "\n",
      "## 121. Mutation Energy of Binding (Pythia-PPI)\n",
      "\n",
      "基于深度学习和多任务学习的预测突变对蛋白-蛋白亲和力影响。\n",
      "\n",
      "## 122. Mutation Energy of Stability\n",
      "\n",
      "Mutation Energy of Stability用于预测突变对蛋白质稳定性的影响。首先能量优化蛋白结构，将一个或多个残基突变为新残基，通过计算折叠自由能变化进行蛋白质稳定性分析。\n",
      "\n",
      "## 123. Mutation Energy of Stability (Pythia)\n",
      "\n",
      "基于自监督图神经网络预测突变对蛋白稳定性影响。\n",
      "\n",
      "## 124. Mutation Energy of Stability (ThermoMPNN)\n",
      "\n",
      "Mutation Energy of Stability (ThermoMPNN)模块基于ThermoMPNN模型实现，此深度神经网络模型可根据蛋白初始结构，预测单点突变对应的稳定性变化。模型使用从ProteinMPNN（一种深度神经网络模型，可根据蛋白质的三维结构预测其氨基酸序列）中提取的结构特征，在已建立的基准数据集上实现了优秀的预测性能。\n",
      "\n",
      "## 125. Mutation Format Conversion\n",
      "\n",
      "Mutation Format Conversion将突变文件中的突变信息加上链名，转换为适用于结构的格式。如将C20S改为CA20S。\n",
      "\n",
      "## 126. Mutation List Generation\n",
      "\n",
      "Mutation List Generation是基于一个原始序列，从经过序列比对后得到的序列（例如BLAST得到的同源序列）中提取每个位点出现过的所有突变（同源突变/共识突变），生成一个突变列表，并按位点统计突变的频率。\n",
      "\n",
      "## 127. Mutation Score (v2.0)\n",
      "\n",
      "Mutation Score是抗体人源化设计中核心模块，是一个基于结构的自动化评分模块。该模块基于抗体的结构信息以及CDR嫁接后的序列信息，对graft后的FR区的每个氨基酸在替换前后的变化程度进行量化评分。评分越高，说明CDR嫁接过程中氨基酸的替换对CDR区的构象可能影响较大，需要进行回复突变。模块输出每个氨基酸的打分值，用于抗体人源化设计流程中后续的分组和人源化抗体序列的生成。\n",
      "\n",
      "## 128. Mutation Score (v2.1)\n",
      "\n",
      "Mutation Score是抗体人源化设计中核心模块，是一个基于结构的自动化评分模块。该模块基于抗体的结构信息以及CDR嫁接后的序列信息，对graft后的FR区的每个氨基酸在替换前后的变化程度进行量化评分。评分越高，说明CDR嫁接过程中氨基酸的替换对CDR区的构象可能影响较大，需要进行回复突变。模块输出每个氨基酸的打分值，用于抗体人源化设计流程中后续的分组和人源化抗体序列的生成。\n",
      "\n",
      "## 129. Mutation Score (v2.3)\n",
      "\n",
      "Mutation Score是抗体人源化设计中核心模块，是一个基于结构的自动化评分模块。该模块基于抗体的结构信息以及CDR嫁接后的序列信息，对graft后的FR区的每个氨基酸在替换前后的变化程度进行量化评分。评分越高，说明CDR嫁接过程中氨基酸的替换对CDR区的构象可能影响较大，需要进行回复突变。模块输出每个氨基酸的打分值，用于抗体人源化设计流程中后续的分组和人源化抗体序列的生成。\n",
      "\n",
      "## 130. Mutation Score (v2.4)\n",
      "\n",
      "Mutation Score是抗体人源化设计中核心模块，是一个基于结构的自动化评分模块。该模块基于抗体的结构信息以及CDR嫁接后的序列信息，对graft后的FR区的每个氨基酸在替换前后的变化程度进行量化评分。评分越高，说明CDR嫁接过程中氨基酸的替换对CDR区的构象可能影响较大，需要进行回复突变。模块输出每个氨基酸的打分值，用于抗体人源化设计流程中后续的分组和人源化抗体序列的生成。\n",
      "\n",
      "## 131. NGS Analysis\n",
      "\n",
      "用于抗体NGS测序的DNA序列分析，具体分析内容包括：IGV、IGD、IGJ基因型标注；DNA序列翻译为氨基酸序列（抗体），并进行CDR识别；基于蛋白（抗体）语言模型（ESM/IgLM），分析不常见残基及优势突变；PTM（翻译后修饰）风险位点分析，标记低、高风险位点；序列特征计算（等电点pI，分子量kDa，疏水性）；序列聚类分析；体系超突变率分析等。\n",
      "\n",
      "## 132. NPT MDP Generation\n",
      "\n",
      "该模块主要是生成等温等压（NPT）的MDP文件，此文件是Gromacs分子动力学模拟需要用到输入文件，里面包含各种参数。\n",
      "\n",
      "## 133. Nanobody Humanization (Llamanade)\n",
      "\n",
      "Nanobody Humanization模块是基于Llamanade实现，Llamanade基于NGS数据库和高分辨率结构，系统分析了Nbs的序列和结构特性，进而确定了可能有助于提高溶解度、结构稳定性和抗原结合的保守残基，以促进Nbs的人源化的理性设计，已成功应用于一批结构多样、强效的SARS-CoV-2中和Nbs人源化工作。对给定的Nbs进行全面人源化分析只需不到一分钟时间。\n",
      "\n",
      "## 134. PAINS Filter\n",
      "\n",
      "通过使用SMARTS模式进行子结构搜索，从输入文件中过滤Filter Pan-assay Interference molecules (PAINS) ，并将适当的分子写入输出文件或仅计算过滤分子的数量。\n",
      "\n",
      "## 135. PDB File\n",
      "\n",
      "PDB文件是一个用于指定PDB文件的模块，可用于其他模块的输入。\n",
      "\n",
      "## 136. PDB Fixer\n",
      "\n",
      "PDB Fixer是修复PDB文件中的问题的模块，包括添加丢失的重原子、添加缺失的氢原子、建立缺失的loop、将非标准残基转化为标准残基、为列出多个替代位置的原子选择一个位置、从模型中删除不需要的链、删除不需要的小分子、为显式溶剂模拟构建一个水盒子。\n",
      "\n",
      "## 137. PDB Insertion Removal\n",
      "\n",
      "PDB Insertion Removal模块用于去掉抗体PDB文件中的插入序列，因为某些计算工具不支持PDB中的插入序列。比如，20A改成20。\n",
      "\n",
      "## 138. PDB Mutation\n",
      "\n",
      "PDB Mutation是用于突变PDB格式的蛋白质结构并返回突变后的结构。\n",
      "\n",
      "## 139. PDB ReNumbering\n",
      "\n",
      "PDB ReNumbering是针对蛋白残基重新编号的工具模块，同时支持抗体kabat，imgt以及chothia的重编号。输入蛋白结构PDB文件，输出重新编号后的PDB文件。\n",
      "\n",
      "## 140. PTM Hotspot by Sequence\n",
      "\n",
      "扫描抗体序列发现潜在的翻译后修饰（PTM）风险位点， PTM 位点是生物制剂开发的常见风险。 通常建议使用WeSeq中的PTM功能进行可视化的分析，本模块更常用于组装自动化流程。\n",
      "\n",
      "## 141. PTM Hotspot by Structure\n",
      "\n",
      "基于结构预测蛋白中高风险的PTM位点，比基于序列的方式更精准。当前版本支持天冬氨酸（ASP）位点发生异构化的概率。\n",
      "\n",
      "## 142. Patch Analysis\n",
      "\n",
      "Patch通常指蛋白质表面的面积较大的正电、负电、疏水残基富集区域，有可能会造成蛋白分子相互作用或聚集等，通过突变将patch中和掉很多情况下可以有效改善聚集程度或粘度。该模块分析蛋白质表面的Patch的大小和分布，用于解决蛋白质的聚集等问题。一般建议通过WeView三维结构可视化编辑器来使用该功能，可以在三维结构中直观地查看patch的位置。\n",
      "\n",
      "## 143. Patent BLAST\n",
      "\n",
      "针对抗体全长或者CDR区进行序列检索的模块。从专利中检索一条抗体可变区时，现有的BLAST程序（例如NCBI BLAST）通常是以全序列进行检索，但是对于抗体而言，功能主要取决于CDR，FR相对不重要，并且由于FR的通用性，许多不同抗体的FR是相同或高度同源的，而FR占序列的比重更高，就导致以抗体的可变区BLAST会得到很多FR相似但CDR不相似的序列。并且，专利申请时，除了保护可变区完整序列，很多情况也会对抗体CDR进行单独保护，以获得更大的保护范围，因此在抗体开发过程中，以CDR为目标进行同源序列检索就很有必要了。为此，唯信团队开发了该程序，可以从现有专利库中检索到与目标CDR最接近的序列。数据更新于：Dec, 2024\n",
      "\n",
      "## 144. Patent CDR BLAST\n",
      "\n",
      "针对抗体CDR区进行序列检索的模块。从专利中检索一条抗体可变区时，现有的BLAST程序（例如NCBI BLAST）通常是以全序列进行检索，但是对于抗体而言，功能主要取决于CDR，FR相对不重要，并且由于FR的通用性，许多不同抗体的FR是相同或高度同源的，而FR占序列的比重更高，就导致以抗体的可变区BLAST会得到很多FR相似但CDR不相似的序列。并且，专利申请时，除了保护可变区完整序列，很多情况也会对抗体CDR进行单独保护，以获得更大的保护范围，因此在抗体开发过程中，以CDR为目标进行同源序列检索就很有必要了。为此，唯信团队开发了该程序，可以从现有专利库中检索到与目标CDR最接近的序列。数据更新于：Jul 20, 2024\n",
      "\n",
      "## 145. Patent Sequence Listing\n",
      "\n",
      "Patent Sequence Listing是批量从专利文本文件中提取序列的工具。很多大分子专利会附带一个序列清单文件，里面存储了专利要求中的全部序列，但是人工很难高效读取，利用此模块可以一次性批量提取。其中Image(OCR)是基于图像的蛋白质序列转换为3个字母编码或1个字母编码的序列。\n",
      "\n",
      "## 146. Peptide Structure Generation\n",
      "\n",
      "Peptide Structure Generation模块是基于多肽序列生成多肽结构：输入多肽的氨基酸序列，生成线性多肽的二维或者三维结构文件，一般用于小肽结构的创建。\n",
      "\n",
      "## 147. Peptide VS\n",
      "\n",
      "Peptide VS模块集成了AutoDock Vina与AutoDock CrankPep进行蛋白-多肽的分子对接，从而预测蛋白-多肽的构象、得到分子对接的能量以及结合亲和力。\n",
      "\n",
      "## 148. Pocket Finder\n",
      "\n",
      "Pocket Finder模块基于几何特性和物理化学特性来识别这些口袋，其主要功能是快速、准确地识别蛋白质表面的潜在口袋。蛋白质口袋（或活性位点）是蛋白质表面的小区域，通常是药物分子或其他小分子结合的地方。识别这些口袋对于药物设计和蛋白质功能研究至关重要。\n",
      "\n",
      "## 149. Property Filter\n",
      "\n",
      "Property Filter模块可以基于导入的分子属性（例如从SDF文件导入）或在运行时对分子进行计算来选择分子的子集。支持的输入文件格式为：SD（.sdf，.sd）。支持的输出文件格式为：SD（.sdf，.sd）。\n",
      "\n",
      "## 150. Protein BLAST\n",
      "\n",
      "Protein BLAST是蛋白Blast数据库，该数据库序列整合了GenPept、Swissprot、PIR、PDF、PDB、RefSeq等序列数据库。\n",
      "\n",
      "## 151. Protein Binder Design (RFDiffusion)\n",
      "\n",
      "扩散模型在图像和语言生成建模方面取得了相当大的成功，但在蛋白质建模方面的应用成功有限，这可能是由于蛋白质主链几何形状和序列-结构关系的复杂性所致。David Baker课题组基于扩散概率模型开发了RFdiffusion方法，通过在蛋白质结构去噪任务上微调RoseTTAFold结构预测网络，获得了一种蛋白质主链生成模型，它在无条件和拓扑约束蛋白质单体设计、蛋白质结合物设计、对称寡聚体设计、酶活性位点支架和对称基序支架等方面取得了出色的性能，用于治疗和金属结合蛋白质设计。RFdiffusion能够从简单的分子规格中设计出多样的、复合的、功能性的蛋白质。\n",
      "\n",
      "## 152. Protein Design (ABACUS-R)\n",
      "\n",
      "ABACUS-R是一种基于深度学习的方法，用于设计能自主地折叠到给定目标骨架的氨基酸序列。该方法通过使用多任务学习训练的编码器-解码器网络，从中心残基的三维局部环境中预测其侧链类型。 若使用该模块发表论文请引用：Liu, Y., Zhang, L., Wang, W. et al. Rotamer-free protein sequence design based on deep learning and self-consistency. Nat Comput Sci. 2022 June 21; 2:451–462.\n",
      "\n",
      "## 153. Protein Design (LigandMPNN)\n",
      "\n",
      "LigandMPNN是一种基于深度学习的蛋白质序列设计方法，专门用于模拟蛋白质与非蛋白质组分（如小分子、核苷酸和金属）之间的相互作用。它是 ProteinMPNN的升级版，能够在蛋白质设计中加入非蛋白质的组分，从而提升对非蛋白-蛋白相互作用的理解。\n",
      "\n",
      "## 154. Protein Design (ProteinMPNN)\n",
      "\n",
      "ProteinMPNN是一种基于深度学习的蛋白质序列设计方法，可基于给定的骨架结构生成合理的序列。在训练过程中加入噪声提高了蛋白质结构模型的序列恢复率，并且产生的序列可以更稳健地编码它们的结构。在天然蛋白质骨架上，ProteinMPNN的序列恢复率为52.4%，高于Rosetta的32.9%。X射线晶体学、低温电镜和功能研究证明了ProteinMPNN的广泛实用性和高准确性，它成功挽救了以前用Rosetta或AlphaFold设计失败的蛋白质单体、环状同源多聚体、四面体纳米颗粒和目标结合蛋白等。本模块也集成了基于ProteinMPNN使用抗体数据微调得到的AbMPNN模型，可更好地进行抗体设计。\n",
      "\n",
      "## 155. Protein Design (RFDesign)\n",
      "\n",
      "Protein Design (RFDesign)是基于RoseTTAFold进行蛋白设计（幻想和修改）的一种相当通用的方法。除了提供需进行设计的功能位点的结构和序列外，无需其他输入，而且与目前的非深度学习方法不同，不需要指定二级结构或骨架的拓扑结构，并能同时产生序列和结构。\n",
      "\n",
      "## 156. Protein Design (RFDiffusion)\n",
      "\n",
      "扩散模型在图像和语言生成建模方面取得了相当大的成功，但在蛋白质建模方面的应用成功有限，这可能是由于蛋白质主链几何形状和序列-结构关系的复杂性所致。David Baker课题组基于扩散概率模型开发了RFdiffusion方法，通过在蛋白质结构去噪任务上微调RoseTTAFold结构预测网络，获得了一种蛋白质主链生成模型，它在无条件和拓扑约束蛋白质单体设计、蛋白质结合物设计、对称寡聚体设计、酶活性位点支架和对称基序支架等方面取得了出色的性能，用于治疗和金属结合蛋白质设计。RFdiffusion能够从简单的分子规格中设计出多样的、复合的、功能性的蛋白质。\n",
      "\n",
      "## 157. Protein Docking (FRODOCK)\n",
      "\n",
      "FRODOCK可以在两个蛋白质结构之间执行详尽的6D对接。这种近似能够非常有效地产生许多关于这两种蛋白质如何相互作用的潜在预测。由于它是刚体近似，因此仅当预期结合后构象变化减少时才有效，通常用作第一步初始对接，后面可以接进一步柔性对接。\n",
      "\n",
      "## 158. Protein Docking (HDOCK)\n",
      "\n",
      "HDOCK是一个集成了同源搜索、基于模板建模、结构预测、大分子对接、生物信息整合的快速蛋白质-蛋白质对接程序。HDOCK使用基于快速傅里叶变换 (FFT) 的对接算法对所有结合模式进行全局采样，然后通过迭代导出的基于知识的评分函数对结合模式进行打分。在多个基准测试中显示很好的预测效果。服务器使用混合对接策略来预测两种分子（如蛋白质和核酸）之间的结合复合物。基于模板的建模和从头开始的自由对接的混合算法进行蛋白质-蛋白质和蛋白质- DNA/RNA 对接。\n",
      "\n",
      "## 159. Protein Isoelectric Point (pI)\n",
      "\n",
      "Protein Isoelectric Point（pI），即分子不带净电荷的pH值，是影响分子理化性质甚至功能的关键参数。该模块使用多种不同的算法，基于序列计算分子的pI数值，并可以对多条链的结果进行合并计算。 基于唯信团队使用部分内部抗体实测pI数据的对比，Sillero算法的精度相对更高，推荐采用。\n",
      "\n",
      "## 160. Protein Physico-chemical Properties\n",
      "\n",
      "Protein Physico-chemical Properties是计算蛋白序列理化性质模块，计算的蛋白理化性质包括：分子质量、等电点、消光系数、不稳定系数、蛋白质的芳香值、总平均亲水性以及二级结构占比。\n",
      "\n",
      "## 161. Protein Protonation\n",
      "\n",
      "蛋白质子化模块主要是预测每个蛋白残基的pKa并根据指定的pH值判断每个残基的质子化状态。\n",
      "\n",
      "## 162. Protein Sequence Generation (ProGen)\n",
      "\n",
      "ProGen是一种语言模型，可以在大型蛋白质家族中生成具有功能的蛋白质序列，类似于在各种话题上生成语法和语义正确的自然语言句子。该模型使用来自>19,000个家族的2.8亿个蛋白质序列进行训练，并附加了控制标签以指定蛋白质属性。可以进一步对ProGen进行微调，以改善来自具有足够同源样本家族的蛋白质生成性能。\n",
      "\n",
      "## 163. Protein Structure Prediction (AlphaFold2)\n",
      "\n",
      "AlphaFold2 是一个高度准确的蛋白质结构预测工具，是目前最高精度的方法之一，甚至接近实验水平。当前模块使用版本：2.1.0。\n",
      "\n",
      "## 164. Protein Structure Prediction (AlphaFold2.3.0)\n",
      "\n",
      "AlphaFold2 是一个高度准确的蛋白质结构预测包，是目前最高精度的方法之一，甚至接近实验水平。模型已进入 CASP14 并发表在 Nature 上。当前使用版本为2.3.0。当前版本已失效，请使用最新版本<a href=\"\"/#/frontend/home/system-workflow-module/update/286\"\">Protein Structure Prediction (AlphaFold2.3.2)。</a>\n",
      "\n",
      "## 165. Protein Structure Prediction (AlphaFold2.3.2)\n",
      "\n",
      "AlphaFold2 是一个高度准确的蛋白质结构预测算法，是目前最高精度的方法之一，在CASP14部分测试中的表现接近实验水平，主要适用于有一定同源序列的蛋白及复合物。 v2.3.2是截止于2023年10月的最新版本。推荐使用AF3 like模块（比如Boltz-1、Chai-1、HelixFold3和Protenix等）。\n",
      "\n",
      "## 166. Protein Structure Prediction (ESMFold)\n",
      "\n",
      "ESMFold是Meta公司开发的蛋白结构预测模型，使用大型语言模型从主序列直接推断结构，预测的速度比AlphaFold方法快60倍，同时能够保持分辨率和准确性。\n",
      "\n",
      "## 167. Protein Structure Prediction (RaptorX-Single)\n",
      "\n",
      "Protein Structure Prediction (RaptorX-Single)模块基于RaptorX-Single算法实现，RaptorX-Single是一种基于单一序列的蛋白质结构预测方法，无需multiple sequence alignment(MSA)信息。它集成了多个蛋白质语言模型和一个结构生成模块，研究结果表明，RaptorX-Single除了比AlphaFold2等基于MSA的方法运行得更快之外，在预测抗体结构、极少同源序列的蛋白和单突变效应方面也优于AlphaFold2和其他无MSA的方法。当预测的蛋白序列有大量同源序列时，RaptorX-Single的预测结果也优于AlphaFold2。\n",
      "\n",
      "## 168. Protein-Protein Interaction (STRING)\n",
      "\n",
      "Protein-Protein Interaction (STRING)是基于STRING的提取蛋白相互作用模块。String是一个蛋白互作网络数据库,包含蛋白直接物理作用的互作关系以及间接作用的互作关系。\n",
      "\n",
      "## 169. RNA 3D Structure Prediction\n",
      "\n",
      "在给定二级结构和实验限制的情况下，从头预测RNA的三维结构模型（可长达约 300 nts ）。除了要预测的 RNA 序列外，您还需要提供一个描述二级结构的文件：具有以圆点符号表示的二级结构的文本文件。\n",
      "\n",
      "## 170. RNA Secondary Structure Prediction\n",
      "\n",
      "使用动态编程算法预测单链RNA或DNA序列的二级结构，返回单一的最佳结构和最低自由能。\n",
      "\n",
      "## 171. Ramachandran Plots\n",
      "\n",
      "Ramachandran Plots模块是对同源建模后模型质量的评估，仅仅考虑蛋白的构象是否合理，并不涉及能量问题。\n",
      "\n",
      "## 172. Receptor-Based Peptide Design\n",
      "\n",
      "Receptor-Based Peptide Design为进行基于受体结构（目前支持单链）的结合多肽设计。\n",
      "\n",
      "## 173. Replace Chain Name\n",
      "\n",
      "Replace Chain Name用于替换PDB文件中的链名。\n",
      "\n",
      "## 174. Restrained Complex Structure Prediction\n",
      "\n",
      "Restrained Complex Structure Prediction模块基于ColabDock框架实现，ColabDock框架通过整合多种实验限制条件，显著提升了蛋白-蛋白对接预测的准确性。\n",
      "\n",
      "## 175. Retrosynthetic Prediction (AiZynthFinder)\n",
      "\n",
      "Retrosynthetic Prediction (AiZynthFinder)是阿斯利康开发的针对小分子的逆反应合成路线预测算法。AiZynthFinder算法基于蒙特卡罗树搜索最终得到可被购买的小分子，用于合成输出分子。树搜索策略采用神经网络方法对已知的反应库进行训练得到。\n",
      "\n",
      "## 176. SD File\n",
      "\n",
      "SD File用于指定SDF格式的小分子结构文件的模块，用于一个文件在多个模块的输入。\n",
      "\n",
      "## 177. SDF File Split\n",
      "\n",
      "SDF File Split是个化合物库文件分割模块，可以将一个大的SDF文件分割为多个SDF文件，支持按文件个数或者分子数目分割，使得分割后的每个SD文件分子数目接近。\n",
      "\n",
      "## 178. SDF Viewer\n",
      "\n",
      "SDF Viewer是小分子化合物库的可视化模块，可以针对一个SDF文件生成一个结构和属性可视、可交互、可检索的HTML页面，方便浏览化合物的结构和属性信息。\n",
      "\n",
      "## 179. Salts Removal\n",
      "\n",
      "Salts Removal模块是从分子中去除盐或者简单地计算含盐分子的数量。\n",
      "\n",
      "## 180. Scaffold Constrained Generation\n",
      "\n",
      "传统分子生成模型无法限制特定骨架，限制了分子生成在结构优化中的应用，Scaffold Constrained Generation骨架限制的生成模型，可以限制骨架，指定优化部位，特异性的生成全新分子库。\n",
      "\n",
      "## 181. SeqKit\n",
      "\n",
      "SeqKit模块是一款超快速、全面的FASTA/Q处理工具包，能够快速完成常见的FASTA/Q文件操作。\n",
      "\n",
      "## 182. Sequence Clustering\n",
      "\n",
      "Sequence Clustering使用DBSCAN算法对多序列比对（MSA）后的结果进行聚类分析，将多序列分为多个cluster类别，并通过可视化模块UMAP进行序列的embedding，并获取二维可视化信息。\n",
      "\n",
      "## 183. Sequence Embedding Generation\n",
      "\n",
      "Sequence Embedding Generation模块基于ESMFold大规模预训练蛋白语言模型实现。提取序列的向量化特征信息（embeddings），可用于下游序列性质（如：突变对应的亲和力变化、稳定性变化，抗体序列可开发性等）预测任务，为判别模型的训练提供序列特征。ESM模型是通用蛋白质语言模型，采用UniRef50/90等序列数据库（数千万条序列）进行模型训练，提供了不同参数量（800万，3500万，1.5亿，6.5亿，30亿，150亿）的各类模型，可用于直接从蛋白序列预测结构、功能和其他蛋白质性质。如在结构预测中，ESM避免了对外部进化数据库、MSA和模板的需求，计算精度与AlphaFold2（存在MSA信息时）接近，无可用MSA信息时，计算精度ESMFold要显著优于AlphaFold2。计算速度比AlphaFold2快数十倍。\n",
      "\n",
      "## 184. Sequence Mutagenesis (Directed)\n",
      "\n",
      "Sequence Mutagenesis (Directed)是根据模板序列批量生成突变体的模块。 用户可以在文本文件中定义所有突变位置和突变氨基酸。\n",
      "\n",
      "## 185. Sequence Mutagenesis (Directed) for Ab\n",
      "\n",
      "Sequence Mutagenesis (Directed) for Ab是根据模板抗体序列和描述突变的突变文件 (json) 批量生成突变抗体序列，通常突变文件由 BLAST 和 MSA 自动生成。这对于高通量抗体工程设计很有用。\n",
      "\n",
      "## 186. Sequence Mutagenesis (Saturated)\n",
      "\n",
      "Sequence Mutagenesis (Saturated)是用于枚举蛋白质序列指定位置饱和突变的所有可能性，生成所有对应突变的文本文件和突变体序列文件。\n",
      "\n",
      "## 187. Sequence Mutation\n",
      "\n",
      "Sequence Mutation是蛋白序列突变模块，用于针对特定位点批量生成突变序列。突变策略包括基于位置的突变，基于同源序列的突变，基于抗体CDR区的突变，以及基于抗体CDR区和同源性的突变。突变类型支持丙氨酸突变，组氨酸突变，以及饱和突变。\n",
      "\n",
      "## 188. Sequence Translation\n",
      "\n",
      "Sequence Translation是DNA序列转换成RNA序列和蛋白序列的工具。\n",
      "\n",
      "## 189. Small Molecule Minimization\n",
      "\n",
      "Small Molecule Minimization是针对小分子结构进行能量最小化优化并得到优化后的3D结构。支持UFF和MMFF两种分子力场，支持SDG, ETDG, KDG, ETKDG四种构象采样方法，用于生成初始3D构象。\n",
      "\n",
      "## 190. Solubility Score\n",
      "\n",
      "基于序列的蛋白溶解度预测。\n",
      "\n",
      "## 191. Solvent Exposure (SASA)\n",
      "\n",
      "基于蛋白质结构（PDB），计算各个残基的溶剂暴露程度（溶液可及化表面积，solvent accessible surface area, SASA）。\n",
      "\n",
      "## 192. Structural Alignment\n",
      "\n",
      "Structural Alignment是对两个蛋白质的三维结构进行叠合的工具。使用BLOSUM62矩阵和Needleman-Wunsch算法在两个序列之间执行全局配对比对，返回叠合后的蛋白结构，同时输出RMSD值。\n",
      "\n",
      "## 193. Structural Alignment (USalign)\n",
      "\n",
      "Structural Alignment (USalign)是基于USalign的结构叠合工具。\n",
      "\n",
      "## 194. Structural Energy\n",
      "\n",
      "基于物理模型（分子力学经验力场）计算多个蛋白结构的能量，并与参考蛋白的结构能量进行比较。\n",
      "\n",
      "## 195. Structure Clustering\n",
      "\n",
      "Structure Clustering是基于分子指纹的小分子结构聚类模块，其采用的聚类方法有Butina或任何其他可用的分层聚类方法。\n",
      "\n",
      "## 196. Structure Evolution\n",
      "\n",
      "Structure Evolution模块基于ESMIF模型实现，ESMIF逆折叠模型旨在根据蛋白质主链原子坐标预测蛋白序列。该模型使用AlphaFold2预测的1200万个蛋白质结构进行训练，包含不变几何输入处理层，随后是一个序列到序列的Transformer，对于在结构上保持不变的主干序列实现51%的本地序列恢复率，对于埋藏残基实现72%的恢复率。该模型还经过跨度屏蔽训练，能够容忍缺失的主链坐标，因此可以预测部分被屏蔽结构的序列。该模块既可以用于亲和力成熟，也可以用于稳定性优化。\n",
      "\n",
      "## 197. Structure Minimization\n",
      "\n",
      "Structure Minimization是结构优化模块，支持氢原子优化、氨基酸侧链优化、整体优化三种方式。\n",
      "\n",
      "## 198. Structure Mutagenesis\n",
      "\n",
      "Structure Mutagenesis模块是从蛋白结构文件得到蛋白的序列信息，然后对指定位点进行饱和突变或者丙氨酸突变，得到包含突变信息的突变文件和突变序列。用于后续其他模块进行结构突变。\n",
      "\n",
      "## 199. Structure Prediction (Boltz-1)\n",
      "\n",
      "Structure Prediction (Boltz-1)基于MIT（麻省理工学院）的Boltz-1算法的AF3 like结构预测模型，Boltz-1是一个开源的深度学习模型，融合了模型架构、速度优化和数据处理方面的创新。在预测生物分子复合物的3D结构方面，它达到了AlphaFold3级别的准确度。Boltz-1在一系列基准测试中表现出与最先进的商业模型相当的性能，为结构生物学中可商业化使用的工具树立了新的标杆。\n",
      "\n",
      "## 200. Structure Prediction (Chai-1)\n",
      "\n",
      "基于Chai Discovery, Inc.（OpenAI投资）的Chai-1算法的AF3 like结构预测模型，支持蛋白、核酸、小分子，金属离子等复合物。\n",
      "\n",
      "## 201. Structure Prediction (HelixFold3)\n",
      "\n",
      "百度螺旋桨PaddleHelix团队研发的HelixFold3，在常规的小分子配体、核酸分子（包括DNA和RNA）以及蛋白质的结构预测精度上已与AlphaFold3相媲美。\n",
      "\n",
      "## 202. Structure Prediction (Protenix)\n",
      "\n",
      "Protenix是字节跳动公司AML AI4Science团队复现的pytorch版本的AlphaFold3模型。以下是ByteDance AML AI4Science团队的主要贡献概要： - 模型性能：将Protenix与现有的模型进行了基准测试。Protenix在不同分子类型的结构预测中表现出强大的性能。作为一个完全开源的模型，它使研究人员能够生成新的预测并对模型进行微调，以满足特定的应用需求。 - 方法：在复现过程中，依据AF3的描述实现了Protenix，并优化了一些模糊步骤，纠正了排版错误，并根据模型行为进行了有针对性的调整。通过分享复现经验，希望支持社区在这些改进的基础上进一步推动该领域的发展。 - 可访问性：已将Protenix开源，提供了模型权重、推理代码和可训练代码供研究用途。\n",
      "\n",
      "## 203. Structure Prediction (RFAA)\n",
      "\n",
      "Biomolecular Structure Prediction (RFAA)模块基于RoseTTAFold All-Atom（RFAA）实现。RFAA依托于RoseTTAFold本身的网络框架，小分子相关信息被合理输入。RFAA在蛋白质结构预测准确性方面与AlphaFold2相媲美，在柔性骨架小分子对接方面同样表现出色，可以对蛋白质以及多个核酸链、小分子复合物结构进行合理预测。\n",
      "\n",
      "## 204. Structure Prediction (WFold)\n",
      "\n",
      "第三方复现的AlphaFold3-like的结构预测模型，与AF3预测结果基本一致，精度超越AF2。\n",
      "\n",
      "## 205. Structure Preparation\n",
      "\n",
      "Structure Preparation是蛋白结构处理模块，用于补全缺失原子和残基，以及蛋白氨基酸残基的质子化判断以及加氢操作。\n",
      "\n",
      "## 206. Structure Relax\n",
      "\n",
      "Structure Relax模块是用于消除晶体结构中的不合理构象，优化点突变设计的结构，以及比较多个不同结构的能量等。通过多次迭代进行氨基酸侧链重排以及能量最小化的计算来搜索给定三维结构的在局部能垒的最优构象。\n",
      "\n",
      "## 207. Substructure Search\n",
      "\n",
      "Substructure Search模块是小分子子结构搜索模块，实现在化合物库中查询出含有特定子结构的分子并输出到SDF文件中。子结构搜索是化学信息学研究中的常用操作，也可以用于虚拟筛选，从小分子商业库中搜索出含有特定功能片段的分子用于后续实验验证。\n",
      "\n",
      "## 208. Synthetic Accessibility Score\n",
      "\n",
      "Synthetic Accessibility Score是一个化合物合成可行性评估指标，反映了化合物是否容易合成。其将小分子合成难易程度用1到10区间数值进行评价，越靠近1表明越容易合成，越靠近10表明合成越困难。SA Score已成为一种普遍使用的指标，可用于预测新化合物的合成可行性，加速化合物筛选和药物发现过程。\n",
      "\n",
      "## 209. TCM Chemical Ingredients\n",
      "\n",
      "TCM Chemical Ingredients是基于中药名称提取中药化学成分的模块。\n",
      "\n",
      "## 210. TCR-pMHC Complex Structure Prediction\n",
      "\n",
      "细胞免疫系统是人体免疫的重要组成部分，它使用 T 细胞受体 (TCR) 识别由主要组织相容性复合体 (MHC) 蛋白呈递的肽形式的抗原蛋白。准确定义TCR的结构基础及其与肽-MHC的结合可以为正常和异常免疫提供重要见解，并有助于指导疫苗和免疫疗法的设计。鉴于实验确定的TCR-肽-MHC结构数量有限，而每个个体内的TCR以及抗原靶标数量巨大，因此需要准确的建模方法。该模块基于TCRmodel2实现，TCRmodel2在AlphaFold基础上对TCR-肽-MHC复合物建模做了优化，与原生AlphaFold和其他基于基准测试的TCR-肽-MHC复合物建模方法相比，其准确度相似或更高，可在30分钟内完成复合物结构预测。\n",
      "\n",
      "## 211. Target Prediction (FastTargetPred)\n",
      "\n",
      "Target Prediction (FastTargetPred)是基于二维相似度的小分子靶点预测模块，活性分子及靶点数据来源于ChEMBL数据库。\n",
      "\n",
      "## 212. Target Prioritization (OpenTargets)\n",
      "\n",
      "Target Prioritization (OpenTargets) 是提取疾病相关靶点蛋白的模块，基于OpenTarget数据库及其疾病-靶点相关性打分方法。\n",
      "\n",
      "## 213. Target-based Cyclic Peptide Design\n",
      "\n",
      "Target-based Cyclic Peptide Design模块基于Evobind2模型实现，Evobind2模型是基于AF2实现的，旨在基于目标蛋白质序列设计新型首尾相接(Head-to-Tail)环肽。通过预测的置信度指标直接选择高亲和力结合体，可以指定的结合位点，并通过异构体评价避免对抗性设计，大大提高了成功率。\n",
      "\n",
      "## 214. Template-guided Structure Prediction\n",
      "\n",
      "Template-guided structure prediction基于自定义的蛋白结构模板，采用colabfold进行蛋白结构预测。\n",
      "\n",
      "## 215. Therapeutic Antibody Profiler\n",
      "\n",
      "Therapeutic Antibody Profiler用于快速对抗体进行打分，评估抗体的成药性。基于抗体可变区的结构，计算CDR区域及其周围的表面疏水性程度、正电荷分布程度、负电荷分布程度、Fv区的重、轻链之间的净电荷失衡程度。\n",
      "\n",
      "## 216. Thermostability Prediction\n",
      "\n",
      "基于TemBERTure开发的Thermostability Prediction是一个用于预测蛋白质热稳定性的深度学习工具，专注于氨基酸序列分析。它包括两个模型：TemBERTureCLS和TemBERTureTm。TemBERTureCLS是一个分类模型，用于预测蛋白质序列的热类别，即判断其是嗜热的还是非嗜热的。TemBERTureTm是一个回归模型，用于根据蛋白质序列预测其熔点温度（Tm）。这两个模型都基于protBERT-BFD语言模型，该模型在大量蛋白质序列数据集上进行了预训练。通过基于适配器的方法进行高效微调，使得TemBERTure能够在不需要广泛重新训练的情况下，稳健地适应特定任务。\n",
      "\n",
      "## 217. Toxic Fragment Identification\n",
      "\n",
      "AlphaTox模块用于识别小分子的毒效片段，从文献中收集了大量的毒效片段构成毒效片段库，利用子结构匹配方法，实现对化合物库中每个分子进行毒效片段匹配，并通过不同颜色区分。\n",
      "\n",
      "## 218. Tumor Gene Expression (TCGA)\n",
      "\n",
      "基于TCGA和GTEx等数据，检索指定基因在肿瘤和正常组织的表达情况，统计并绘制肿瘤细胞、肿瘤组织、正常组织等的基因表达差异，帮助药物靶点选择、研发立项和决策。\n",
      "\n",
      "## 219. Venn Diagram Plot\n",
      "\n",
      "Venn Diagram Plot是一个制作韦恩图(Venn diagram)模块，常用于比较两个集合的重叠区域以及提取公共部分内容。例如，用于中药网络药理学分析中提取中药成分预测靶点与疾病相关靶点的交集。\n",
      "\n",
      "## 220. mRNA 5'UTRs optimization\n",
      "\n",
      "mRNA 5'UTRs optimization基于Smart5UTR模型实现，Smart5UTR 是一种新颖的深度生成模型，设计用于在 mRNA 序列中创建 N1-甲基假尿苷 （m1Ψ） 5'UTR。Smart5UTR 利用多任务自动编码器框架，利用从大型数据集中学习到的潜在特征，有效地生成 5'UTR 序列。Smart5UTR设计的mRNA的性能已通过体外和体内实验得到验证。这个强大的工具简化了m1Ψ-5'UTRs的设计，有助于开发更有效的mRNA疗法。\n",
      "\n",
      "## 221. mRNA Optimization (AlphaRNA)\n",
      "\n",
      "mRNA Optimization (AlphaRNA)可以优化mRNA序列以获得更好的密码子偏好性和更稳定的二级结构，以优化其表达量和半衰期、抗体滴度等。支持人、小鼠、大鼠、猪等种属。 在输入UTR的情况下，若不选择固定UTR，则会同时优化UTR与CDS；若选择固定UTR，则只优化CDS区，但CDS区域仍会尽量与UTR配对。 若不希望CDS与UTR配对，只优化CDS自身，则不需要输入UTR，只输入CDS即可。\n",
      "\n",
      "## 222. siRNA Designer\n",
      "\n",
      "siRNA Designer基于靶点基因序列，设计siRNA分子序列。\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with open(\"cloud_modules_en.md\") as f:\n",
    "    doc_en_md = f.read()\n",
    "print(doc_en_md)\n",
    "\n",
    "with open(\"cloud_modules_cn.md\") as f:\n",
    "    doc_cn_md = f.read()\n",
    "print(doc_cn_md)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dac74822-f6f8-4920-bc2c-123889e5dee4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Embedding time =  0.8139917850494385\n",
      "Retrieval time =  0.839777946472168\n",
      "\n",
      "Answer:\n",
      "\n",
      "Modify the configuration files as shown below:\n",
      "\n",
      "config_setting_1 = true\n",
      "config_setting_2 = false\n"
     ]
    }
   ],
   "source": [
    "from utils.open_rag import *\n",
    "        \n",
    "# Example usage\n",
    "markdown_text = \"\"\"\n",
    "# Introduction\n",
    "\n",
    "Welcome to our guide.\n",
    "\n",
    "Here we will cover several important aspects.\n",
    "\n",
    "<pre>\n",
    "Author    James Lee\n",
    "          ABC Inc,\n",
    "          USA\n",
    "\n",
    "April 25, 2024\n",
    "</pre>\n",
    "\n",
    "## Setup Instructions\n",
    "\n",
    "First, ensure you have the following tools installed:\n",
    "\n",
    "- Tool A\n",
    "- Tool B\n",
    "\n",
    "Please follow these steps to get started.\n",
    "\n",
    "## Configuration\n",
    "\n",
    "Modify the configuration files as shown below:\n",
    "\n",
    "```\n",
    "config_setting_1 = true\n",
    "config_setting_2 = false\n",
    "```\n",
    "\n",
    "Remember to restart the service after changing the config files.\n",
    "\n",
    "## Data Format\n",
    "\n",
    "Our system uses the following data structure:\n",
    "\n",
    "Table 1: Table of names and their values.\n",
    "| ID | Name   | Value |\n",
    "|----|--------|-------|\n",
    "| 1  | Item 1 | 100   |\n",
    "| 2  | Item 2 | 200   |\n",
    "Note: more data will be added.\n",
    "\n",
    "Please make sure your data conforms to this table.\n",
    "\n",
    "# Summary\n",
    "\n",
    "This guide should help you get started with the basic setup and configuration.\n",
    "\n",
    "For more details, visit our [website](http://example.com).\n",
    "\n",
    "Thank you for reading!\n",
    "\"\"\"\n",
    "\n",
    "# Create Q&A agent\n",
    "qna_agent = QnAAgent(markdown_text, embedding_function=get_embedding_openai)\n",
    "\n",
    "# Example of asking a question\n",
    "question = \"How do I configure the settings?\"\n",
    "answer = qna_agent.answer_question(question)\n",
    "print(\"\\nAnswer:\\n\")\n",
    "print(answer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a760cac5-d6ab-4613-b0b0-25ea23148de3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time A =  15.33266305923462\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'doc_md' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 6\u001b[0m\n\u001b[1;32m      4\u001b[0m time2 \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTime A = \u001b[39m\u001b[38;5;124m\"\u001b[39m, time2\u001b[38;5;241m-\u001b[39mtime1)\n\u001b[0;32m----> 6\u001b[0m qna_agent2 \u001b[38;5;241m=\u001b[39m QnAAgent(\u001b[43mdoc_md\u001b[49m,embedding_function\u001b[38;5;241m=\u001b[39mget_embedding_openai)\n\u001b[1;32m      7\u001b[0m time3 \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTime B = \u001b[39m\u001b[38;5;124m\"\u001b[39m, time3\u001b[38;5;241m-\u001b[39mtime2)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'doc_md' is not defined"
     ]
    }
   ],
   "source": [
    "# Create Q&A agent\n",
    "time1 = time.time()\n",
    "qna_agent = QnAAgent(markdown_text,embedding_function=get_embedding_openai)\n",
    "time2 = time.time()\n",
    "print(\"Time A = \", time2-time1)\n",
    "qna_agent2 = QnAAgent(doc_md,embedding_function=get_embedding_openai)\n",
    "time3 = time.time()\n",
    "print(\"Time B = \", time3-time2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e585972-de84-4489-ad0b-9e3cc51da096",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example of asking a question\n",
    "time1 = time.time()\n",
    "question = \"How do I configure the settings?\"\n",
    "answer = qna_agent.answer_question(question)\n",
    "print(\"\\nAnswer:\\n\")\n",
    "print(answer)\n",
    "\n",
    "question = \"What kind of tools need to be installed according to the instruction?\"\n",
    "answer = qna_agent.answer_question(question)\n",
    "print(\"\\nAnswer:\\n\")\n",
    "print(answer)\n",
    "\n",
    "print(\"Time = \", time.time()-time1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d73a10a-4362-45e2-a38f-ac540bf17345",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4946f8b0-99cf-4755-b8fe-92bb9a0d6c54",
   "metadata": {},
   "outputs": [],
   "source": [
    "time1 = time.time()\n",
    "query = \"What is ChatMol?\"\n",
    "answer = qna_agent2.answer_question(query,show_content=True)\n",
    "print(\"--------------------------------------\")\n",
    "print(\"Q:\",query)\n",
    "print(\"A:\",answer)\n",
    "time2 = time.time()\n",
    "\n",
    "print(\"Time = \", time2-time1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b91cc3f7-debb-4658-8b8f-dcf77e20c9e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "time1 = time.time()\n",
    "query = \"Show the original abstract of this paper\"\n",
    "answer = qna_agent2.answer_question(query,show_content=True)\n",
    "print(\"--------------------------------------\")\n",
    "print(\"Q:\",query)\n",
    "print(\"A:\",answer)\n",
    "time2 = time.time()\n",
    "\n",
    "print(\"Time = \", time2-time1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2964874e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from openai import OpenAI\n",
    "\n",
    "client = OpenAI()\n",
    "\n",
    "def get_embedding(sentences, model=\"text-embedding-3-small\"):\n",
    "   data = client.embeddings.create(input = sentences, model=model).data\n",
    "   vecs = np.array([data[i].embedding for i in range(len(data))])\n",
    "   return vecs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "224cbd0c-ea85-42d5-876d-16c450748c1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ex = get_embedding(['This is a test','The second test'])\n",
    "print(ex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99bedf1f-6f36-42c0-aed1-222d58da110d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from openai import OpenAI\n",
    "import time\n",
    "\n",
    "time1 = time.time()\n",
    "client = OpenAI(\n",
    "    base_url='http://localhost:11434/v1/',\n",
    "    #base_url = 'http://100.89.180.132:11434/v1',\n",
    "    #base_url = 'https://www.chatmol.org/ollama/v1/',\n",
    ")\n",
    "\n",
    "ChatMol_md = doc2_md\n",
    "\n",
    "prompt_science_paper = \"Please read carefully the following scientific paper in markdown format, then answer a user question based on the information in this paper.\"\n",
    "\n",
    "model = 'llama3.2-3B-16k'\n",
    "messages=[\n",
    "        {\"role\": \"system\",\"content\": prompt_science_paper},\n",
    "        {\"role\": \"user\",\"content\": f\"Here is the paper in markdown format: \\n\\n{ChatMol_md}\\n\\n\"},\n",
    "        {\"role\": \"user\",\"content\": \"What is SAFE?\"}\n",
    "    ]\n",
    "response = client.chat.completions.create(\n",
    "    messages=messages,\n",
    "    model=model,\n",
    ")\n",
    "print(response.choices[0].message.content)\n",
    "print(\"Time = \", time.time()-time1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1891cba0-b16b-48f9-abf8-4be77312d65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "call_llm_api(\"Hello!\", model_name='llama3.2-3B-16k', api_url='http://localhost:11434/api/generate',temperature=0.1, max_tokens=1024)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad465e68-a473-40a4-9b44-84e6083ab514",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Set the API key\n",
    "client = OpenAI()\n",
    "\n",
    "MODEL=\"gpt-4o\"\n",
    "\n",
    "completion = client.chat.completions.create(\n",
    "  model=MODEL,\n",
    "  messages=[\n",
    "    {\"role\": \"system\", \"content\": \"You are a helpful assistant that helps me with my math homework!\"},\n",
    "    {\"role\": \"user\", \"content\": \"Hello! Could you solve 20 x 5?\"}\n",
    "  ]\n",
    ")\n",
    "print(\"Assistant: \" + completion.choices[0].message.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "74a4fef5-5485-4050-8408-579e7450bf8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<think>\n",
      "\n",
      "</think>\n",
      "\n",
      "Hello! How can I assist you today? 😊\n",
      "Time =  8.320701837539673\n"
     ]
    }
   ],
   "source": [
    "from openai import OpenAI\n",
    "import time\n",
    "\n",
    "time1 = time.time()\n",
    "client = OpenAI(\n",
    "    #base_url='http://localhost:11434/v1/',\n",
    "    base_url = 'http://100.89.180.132:11434/v1',\n",
    "    #base_url = 'https://www.chatmol.org/ollama/v1/',\n",
    ")\n",
    "\n",
    "model = 'deepseek-r1:70b'\n",
    "messages=[\n",
    "        {\"role\": \"user\",\"content\": \"Hello\"}\n",
    "    ]\n",
    "response = client.chat.completions.create(\n",
    "    messages=messages,\n",
    "    model=model,\n",
    ")\n",
    "print(response.choices[0].message.content)\n",
    "print(\"Time = \", time.time()-time1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b6707f73-40ba-446c-864e-407439baa0c5",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValidationError",
     "evalue": "2 validation errors for EmbedRequest\ninput.str\n  Input should be a valid string [type=string_type, input_value=[['Llamas are members of ...f the camelid family.']], input_type=list]\n    For further information visit https://errors.pydantic.dev/2.10/v/string_type\ninput.json-or-python[json=list[str],python=chain[is-instance[Sequence],function-wrap[sequence_validator()]]].0\n  Input should be a valid string [type=string_type, input_value=['Llamas are members of t...of the camelid family.'], input_type=list]\n    For further information visit https://errors.pydantic.dev/2.10/v/string_type",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValidationError\u001b[0m                           Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 20\u001b[0m\n\u001b[1;32m     18\u001b[0m time1 \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[1;32m     19\u001b[0m \u001b[38;5;66;03m# Generate the embedding for the text\u001b[39;00m\n\u001b[0;32m---> 20\u001b[0m vector \u001b[38;5;241m=\u001b[39m \u001b[43membeddings\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43membed_query\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43mtext\u001b[49m\u001b[43m,\u001b[49m\u001b[43mtext\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     22\u001b[0m \u001b[38;5;66;03m# Display the first 10 dimensions of the embedding vector\u001b[39;00m\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;28mlen\u001b[39m(vector))\n",
      "File \u001b[0;32m/opt/anaconda3/envs/GenAI310/lib/python3.10/site-packages/langchain_ollama/embeddings.py:244\u001b[0m, in \u001b[0;36mOllamaEmbeddings.embed_query\u001b[0;34m(self, text)\u001b[0m\n\u001b[1;32m    242\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21membed_query\u001b[39m(\u001b[38;5;28mself\u001b[39m, text: \u001b[38;5;28mstr\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m List[\u001b[38;5;28mfloat\u001b[39m]:\n\u001b[1;32m    243\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Embed query text.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 244\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43membed_documents\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43mtext\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m[\u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[0;32m/opt/anaconda3/envs/GenAI310/lib/python3.10/site-packages/langchain_ollama/embeddings.py:237\u001b[0m, in \u001b[0;36mOllamaEmbeddings.embed_documents\u001b[0;34m(self, texts)\u001b[0m\n\u001b[1;32m    235\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21membed_documents\u001b[39m(\u001b[38;5;28mself\u001b[39m, texts: List[\u001b[38;5;28mstr\u001b[39m]) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m List[List[\u001b[38;5;28mfloat\u001b[39m]]:\n\u001b[1;32m    236\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Embed search docs.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 237\u001b[0m     embedded_docs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_client\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43membed\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    238\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtexts\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_default_params\u001b[49m\n\u001b[1;32m    239\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124membeddings\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m    240\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m embedded_docs\n",
      "File \u001b[0;32m/opt/anaconda3/envs/GenAI310/lib/python3.10/site-packages/ollama/_client.py:361\u001b[0m, in \u001b[0;36mClient.embed\u001b[0;34m(self, model, input, truncate, options, keep_alive)\u001b[0m\n\u001b[1;32m    349\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21membed\u001b[39m(\n\u001b[1;32m    350\u001b[0m   \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    351\u001b[0m   model: \u001b[38;5;28mstr\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    355\u001b[0m   keep_alive: Optional[Union[\u001b[38;5;28mfloat\u001b[39m, \u001b[38;5;28mstr\u001b[39m]] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    356\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m EmbedResponse:\n\u001b[1;32m    357\u001b[0m   \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_request(\n\u001b[1;32m    358\u001b[0m     EmbedResponse,\n\u001b[1;32m    359\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPOST\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m    360\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/api/embed\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[0;32m--> 361\u001b[0m     json\u001b[38;5;241m=\u001b[39m\u001b[43mEmbedRequest\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    362\u001b[0m \u001b[43m      \u001b[49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    363\u001b[0m \u001b[43m      \u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    364\u001b[0m \u001b[43m      \u001b[49m\u001b[43mtruncate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtruncate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    365\u001b[0m \u001b[43m      \u001b[49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    366\u001b[0m \u001b[43m      \u001b[49m\u001b[43mkeep_alive\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkeep_alive\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    367\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mmodel_dump(exclude_none\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m),\n\u001b[1;32m    368\u001b[0m   )\n",
      "File \u001b[0;32m/opt/anaconda3/envs/GenAI310/lib/python3.10/site-packages/pydantic/main.py:214\u001b[0m, in \u001b[0;36mBaseModel.__init__\u001b[0;34m(self, **data)\u001b[0m\n\u001b[1;32m    212\u001b[0m \u001b[38;5;66;03m# `__tracebackhide__` tells pytest and some other tools to omit this function from tracebacks\u001b[39;00m\n\u001b[1;32m    213\u001b[0m __tracebackhide__ \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m--> 214\u001b[0m validated_self \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m__pydantic_validator__\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalidate_python\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mself_instance\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    215\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m validated_self:\n\u001b[1;32m    216\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    217\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA custom validator is returning a value other than `self`.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m    218\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mReturning anything other than `self` from a top level model validator isn\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt supported when validating via `__init__`.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    219\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSee the `model_validator` docs (https://docs.pydantic.dev/latest/concepts/validators/#model-validators) for more details.\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m    220\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2\u001b[39m,\n\u001b[1;32m    221\u001b[0m     )\n",
      "\u001b[0;31mValidationError\u001b[0m: 2 validation errors for EmbedRequest\ninput.str\n  Input should be a valid string [type=string_type, input_value=[['Llamas are members of ...f the camelid family.']], input_type=list]\n    For further information visit https://errors.pydantic.dev/2.10/v/string_type\ninput.json-or-python[json=list[str],python=chain[is-instance[Sequence],function-wrap[sequence_validator()]]].0\n  Input should be a valid string [type=string_type, input_value=['Llamas are members of t...of the camelid family.'], input_type=list]\n    For further information visit https://errors.pydantic.dev/2.10/v/string_type"
     ]
    }
   ],
   "source": [
    "import time\n",
    "\n",
    "from langchain_ollama import OllamaEmbeddings\n",
    "\n",
    "# Replace with the base URL of your remote Ollama API\n",
    "remote_base_url = \"http://100.89.180.132:11434\"\n",
    "#remote_base_url = \"https://chatmol.org/ollama\"\n",
    "\n",
    "# Initialize the OllamaEmbeddings with the remote base URL and bge-large model\n",
    "embeddings = OllamaEmbeddings(\n",
    "    model=\"bge-large\",\n",
    "    base_url=remote_base_url\n",
    ")\n",
    "\n",
    "# Example text to embed\n",
    "text = \"Llamas are members of the camelid family.\"\n",
    "\n",
    "time1 = time.time()\n",
    "# Generate the embedding for the text\n",
    "vector = embeddings.embed_query([text,text])\n",
    "\n",
    "# Display the first 10 dimensions of the embedding vector\n",
    "print(len(vector))\n",
    "print(vector[:10])\n",
    "print(\"Time = \", time.time()-time1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07f78b28-1f5f-4657-a536-4c915cea661c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
